// @varlet/ui/lib/app-bar/AppBar.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _props = require("./props");

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-app-bar__left"
};
var _hoisted_2 = {
  key: 0,
  class: "var-app-bar__title"
};
var _hoisted_3 = {
  class: "var-app-bar__right"
};

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-app-bar", {
      'var-elevation--3': _ctx.elevation
    }]),
    style: (0, _vue.normalizeStyle)({
      background: _ctx.color,
      color: _ctx.textColor
    })
  }, [(0, _vue.createElementVNode)("div", _hoisted_1, [(0, _vue.renderSlot)(_ctx.$slots, "left"), _ctx.titlePosition === 'left' ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: "var-app-bar__title",
    style: (0, _vue.normalizeStyle)({
      paddingLeft: _ctx.$slots.left ? 0 : undefined
    })
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.title), 1
  /* TEXT */
  )])], 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true)]), _ctx.titlePosition === 'center' ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_2, [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.title), 1
  /* TEXT */
  )])])) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("div", _hoisted_3, [_ctx.titlePosition === 'right' ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: "var-app-bar__title",
    style: (0, _vue.normalizeStyle)({
      paddingRight: _ctx.$slots.right ? 0 : undefined
    })
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.title), 1
  /* TEXT */
  )])], 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.renderSlot)(_ctx.$slots, "right")])], 6
  /* CLASS, STYLE */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarAppBar',
  props: _props.props
});

exports.default = _default;
