// @varlet/ui/es/pagination/Pagination.js
import VarMenu from '../menu';
import Ripple from '../ripple';
import VarIcon from '../icon';
import VarCell from '../cell';
import VarInput from '../input';
import { defineComponent, ref, computed, watch } from 'vue';
import { props } from './porps';
import { isNumber, toNumber } from '../utils/shared';
import { pack } from '../locale';
import { renderSlot as _renderSlot, resolveComponent as _resolveComponent, createVNode as _createVNode, normalizeClass as _normalizeClass, resolveDirective as _resolveDirective, createElementVNode as _createElementVNode, withDirectives as _withDirectives, withKeys as _withKeys, toDisplayString as _toDisplayString, openBlock as _openBlock, createElementBlock as _createElementBlock, createCommentVNode as _createCommentVNode, renderList as _renderList, Fragment as _Fragment, createTextVNode as _createTextVNode, withCtx as _withCtx, createBlock as _createBlock, pushScopeId as _pushScopeId, popScopeId as _popScopeId } from "vue";

var _withScopeId = n => (_pushScopeId(""), n = n(), _popScopeId(), n);

var _hoisted_1 = {
  class: "var-pagination"
};
var _hoisted_2 = ["item-mode", "onClick"];
var _hoisted_3 = {
  key: 4,
  class: "var-pagination__total"
};
export function render(_ctx, _cache) {
  var _component_var_icon = _resolveComponent("var-icon");

  var _component_var_input = _resolveComponent("var-input");

  var _component_var_cell = _resolveComponent("var-cell");

  var _component_var_menu = _resolveComponent("var-menu");

  var _directive_ripple = _resolveDirective("ripple");

  return _openBlock(), _createElementBlock("ul", _hoisted_1, [_withDirectives(_createElementVNode("li", {
    class: _normalizeClass(["var-pagination__item var-pagination__prev", {
      'var-pagination__item-disabled': _ctx.current <= 1 || _ctx.disabled,
      'var-pagination__item-hover': _ctx.simple,
      'var-elevation--2': !_ctx.simple
    }]),
    onClick: _cache[0] || (_cache[0] = $event => _ctx.clickItem('prev'))
  }, [_renderSlot(_ctx.$slots, "prev", {}, () => [_createVNode(_component_var_icon, {
    name: "chevron-left"
  })])], 2
  /* CLASS */
  ), [[_directive_ripple, {
    disabled: _ctx.current <= 1 || _ctx.disabled
  }]]), _ctx.simple ? (_openBlock(), _createElementBlock("li", {
    key: 0,
    class: _normalizeClass(["var-pagination__simple", {
      'var-pagination__item-disabled': _ctx.disabled
    }])
  }, [_createVNode(_component_var_input, {
    modelValue: _ctx.simpleValue,
    "onUpdate:modelValue": _cache[1] || (_cache[1] = $event => _ctx.simpleValue = $event),
    disabled: _ctx.disabled,
    "var-pagination-cover": "",
    onBlur: _cache[2] || (_cache[2] = $event => _ctx.setPage('simple', _ctx.simpleValue, $event)),
    onKeydown: _cache[3] || (_cache[3] = _withKeys($event => _ctx.setPage('simple', _ctx.simpleValue, $event), ["enter"]))
  }, null, 8
  /* PROPS */
  , ["modelValue", "disabled"]), _createElementVNode("span", null, "/ " + _toDisplayString(_ctx.pageCount), 1
  /* TEXT */
  )], 2
  /* CLASS */
  )) : (_openBlock(true), _createElementBlock(_Fragment, {
    key: 1
  }, _renderList(_ctx.pageList, (item, index) => {
    return _withDirectives((_openBlock(), _createElementBlock("li", {
      key: _ctx.toNumber(item) + index,
      "item-mode": _ctx.getMode(item, index),
      class: _normalizeClass(["var-pagination__item var-elevation--2", {
        'var-pagination__item-active': item === _ctx.current && !_ctx.disabled,
        'var-pagination__item-hide': _ctx.isHideEllipsis(item, index),
        'var-pagination__item-disabled': _ctx.disabled,
        'var-pagination__item-disabled-active': item === _ctx.current && _ctx.disabled
      }]),
      onClick: $event => _ctx.clickItem(item, index)
    }, [_createTextVNode(_toDisplayString(item), 1
    /* TEXT */
    )], 10
    /* CLASS, PROPS */
    , _hoisted_2)), [[_directive_ripple, {
      disabled: _ctx.disabled
    }]]);
  }), 128
  /* KEYED_FRAGMENT */
  )), _withDirectives(_createElementVNode("li", {
    class: _normalizeClass(["var-pagination__item var-pagination__next", {
      'var-pagination__item-disabled': _ctx.current >= _ctx.pageCount || _ctx.disabled,
      'var-pagination__item-hover': _ctx.simple,
      'var-elevation--2': !_ctx.simple
    }]),
    onClick: _cache[4] || (_cache[4] = $event => _ctx.clickItem('next'))
  }, [_renderSlot(_ctx.$slots, "next", {}, () => [_createVNode(_component_var_icon, {
    name: "chevron-right"
  })])], 2
  /* CLASS */
  ), [[_directive_ripple, {
    disabled: _ctx.current >= _ctx.pageCount || _ctx.disabled
  }]]), _ctx.showSizeChanger ? (_openBlock(), _createElementBlock("li", {
    key: 2,
    class: _normalizeClass(["var-pagination__size", {
      'var-pagination__item-disabled': _ctx.disabled
    }])
  }, [_createVNode(_component_var_menu, {
    show: _ctx.menuVisible,
    "onUpdate:show": _cache[6] || (_cache[6] = $event => _ctx.menuVisible = $event),
    "offset-x": -4
  }, {
    menu: _withCtx(() => [(_openBlock(true), _createElementBlock(_Fragment, null, _renderList(_ctx.sizeOption, (option, index) => {
      return _withDirectives((_openBlock(), _createBlock(_component_var_cell, {
        class: _normalizeClass(["var-pagination__list", {
          'var-pagination__list-active': _ctx.size === option
        }]),
        key: index,
        onClick: $event => _ctx.clickSize(option)
      }, {
        default: _withCtx(() => [_createTextVNode(_toDisplayString(option) + _toDisplayString(_ctx.pack.paginationItem) + " / " + _toDisplayString(_ctx.pack.paginationPage), 1
        /* TEXT */
        )]),
        _: 2
        /* DYNAMIC */

      }, 1032
      /* PROPS, DYNAMIC_SLOTS */
      , ["class", "onClick"])), [[_directive_ripple]]);
    }), 128
    /* KEYED_FRAGMENT */
    ))]),
    default: _withCtx(() => [_createElementVNode("div", {
      class: "var-pagination__size-open",
      style: {
        "display": "flex"
      },
      onClick: _cache[5] || (_cache[5] = (...args) => _ctx.showMenu && _ctx.showMenu(...args))
    }, [_createElementVNode("span", null, _toDisplayString(_ctx.size) + _toDisplayString(_ctx.pack.paginationItem) + " / " + _toDisplayString(_ctx.pack.paginationPage), 1
    /* TEXT */
    ), _createVNode(_component_var_icon, {
      class: "var-pagination__size-open-icon",
      "var-pagination-cover": "",
      name: "menu-down"
    })])]),
    _: 1
    /* STABLE */

  }, 8
  /* PROPS */
  , ["show"])], 2
  /* CLASS */
  )) : _createCommentVNode("v-if", true), _ctx.showQuickJumper && !_ctx.simple ? (_openBlock(), _createElementBlock("li", {
    key: 3,
    class: _normalizeClass(["var-pagination__quickly", {
      'var-pagination__item-disabled': _ctx.disabled
    }])
  }, [_createTextVNode(_toDisplayString(_ctx.pack.paginationJump) + " ", 1
  /* TEXT */
  ), _createVNode(_component_var_input, {
    modelValue: _ctx.inputValue,
    "onUpdate:modelValue": _cache[7] || (_cache[7] = $event => _ctx.inputValue = $event),
    disabled: _ctx.disabled,
    "var-pagination-cover": "",
    onBlur: _cache[8] || (_cache[8] = $event => _ctx.setPage('quick', _ctx.inputValue, $event)),
    onKeydown: _cache[9] || (_cache[9] = _withKeys($event => _ctx.setPage('quick', _ctx.inputValue, $event), ["enter"]))
  }, null, 8
  /* PROPS */
  , ["modelValue", "disabled"])], 2
  /* CLASS */
  )) : _createCommentVNode("v-if", true), _ctx.totalText ? (_openBlock(), _createElementBlock("li", _hoisted_3, _toDisplayString(_ctx.totalText), 1
  /* TEXT */
  )) : _createCommentVNode("v-if", true)]);
}
export default defineComponent({
  render,
  name: 'VarPagination',
  components: {
    VarMenu,
    VarIcon,
    VarCell,
    VarInput
  },
  directives: {
    Ripple
  },
  props,

  setup(props) {
    var menuVisible = ref(false);
    var inputValue = ref('');
    var simpleValue = ref('1');
    var isHideEllipsisHead = ref(false);
    var isHideEllipsisTail = ref(false);
    var current = ref(toNumber(props.current) || 1);
    var size = ref(toNumber(props.size) || 10);
    var pageList = ref([]);
    var activePosition = computed(() => Math.ceil(props.maxPagerCount / 2));
    var pageCount = computed(() => Math.ceil(toNumber(props.total) / toNumber(size.value)));
    var range = computed(() => {
      var start = size.value * (current.value - 1) + 1;
      var end = Math.min(size.value * current.value, toNumber(props.total));
      return [start, end];
    });
    var totalText = computed(() => {
      if (!props.showTotal) return '';
      return props.showTotal(toNumber(props.total), range.value);
    });

    var isHideEllipsis = (item, index) => {
      if (isNumber(item)) return false;
      return index === 1 ? isHideEllipsisHead.value : isHideEllipsisTail.value;
    };

    var getMode = (item, index) => {
      if (isNumber(item)) return 'basic';
      return index === 1 ? 'head' : 'tail';
    };

    var clickItem = (item, index) => {
      if (item === current.value || props.disabled) return;
      if (isNumber(item)) current.value = item;else if (item === 'prev') current.value > 1 && (current.value -= 1);else if (item === 'next') current.value < pageCount.value && (current.value += 1);else if (item === '...') {
        if (index === 1) {
          current.value = Math.max(current.value - props.maxPagerCount, 1);
        } else {
          current.value = Math.min(current.value + props.maxPagerCount, pageCount.value);
        }
      }
    };

    var showMenu = () => {
      if (props.disabled) return;
      menuVisible.value = true;
    };

    var clickSize = option => {
      size.value = option;
      menuVisible.value = false;
    };

    var isValidatePage = value => {
      var pattern = /^[1-9][0-9]*$/;
      return pattern.test(value);
    };

    var setPage = (type, value, event) => {
      ;
      event.target.blur();

      if (isValidatePage(value)) {
        var valueNum = toNumber(value);

        if (valueNum > pageCount.value) {
          valueNum = pageCount.value;
          simpleValue.value = "" + valueNum;
        }

        if (valueNum !== current.value) current.value = valueNum;
      }

      if (type === 'quick') inputValue.value = '';
      if (type === 'simple' && !isValidatePage(value)) simpleValue.value = "" + current.value;
    };

    watch([() => props.current, () => props.size], ([newCurrent, newSize]) => {
      current.value = toNumber(newCurrent) || 1;
      size.value = toNumber(newSize || 10);
    });
    watch([current, pageCount], ([newCurrent, newCount], [oldCurrent, oldCount]) => {
      if (newCurrent > newCount) {
        current.value = newCount;
        return;
      }

      var list = [];
      var {
        maxPagerCount
      } = props;
      var rEllipseSign = newCount - (maxPagerCount - activePosition.value) - 1;
      simpleValue.value = "" + newCurrent;

      if (newCount - 2 > maxPagerCount) {
        if (oldCurrent === undefined || newCount !== oldCount) {
          for (var i = 2; i < maxPagerCount + 2; i++) {
            list.push(i);
          }
        } // 左边不需要出现省略符号占位


        if (newCurrent <= maxPagerCount && newCurrent < rEllipseSign) {
          list = [];

          for (var _i = 1; _i < maxPagerCount + 1; _i++) {
            list.push(_i + 1);
          }

          isHideEllipsisHead.value = true;
          isHideEllipsisTail.value = false;
        } // 两边都需要出现省略符号占位


        if (newCurrent > maxPagerCount && newCurrent < rEllipseSign) {
          list = [];

          for (var _i2 = 1; _i2 < maxPagerCount + 1; _i2++) {
            list.push(newCurrent + _i2 - activePosition.value);
          } // 针对 maxPagerCount===1 的特殊处理


          isHideEllipsisHead.value = newCurrent === 2 && maxPagerCount === 1;
          isHideEllipsisTail.value = false;
        } // 右边不需要出现省略符号占位


        if (newCurrent >= rEllipseSign) {
          list = [];

          for (var _i3 = 1; _i3 < maxPagerCount + 1; _i3++) {
            list.push(newCount - (maxPagerCount - _i3) - 1);
          }

          isHideEllipsisHead.value = false;
          isHideEllipsisTail.value = true;
        }

        list = [1, '...', ...list, '...', newCount];
      } else {
        for (var _i4 = 1; _i4 <= newCount; _i4++) {
          list.push(_i4);
        }
      }

      pageList.value = list;
      if (oldCurrent !== undefined) props.onChange == null ? void 0 : props.onChange(newCurrent, size.value);
    }, {
      immediate: true
    });
    return {
      pack,
      current,
      menuVisible,
      size,
      pageCount,
      pageList,
      inputValue,
      simpleValue,
      totalText,
      getMode,
      isHideEllipsis,
      clickItem,
      showMenu,
      clickSize,
      setPage,
      toNumber
    };
  }

});
