// @varlet/ui/lib/swipe-item/SwipeItem.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _provide = require("./provide");

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: "var-swipe-item",
    style: (0, _vue.normalizeStyle)({
      width: !_ctx.vertical ? _ctx.size + "px" : undefined,
      height: _ctx.vertical ? _ctx.size + "px" : undefined,
      transform: "translate" + (_ctx.vertical ? 'Y' : 'X') + "(" + _ctx.translate + "px)"
    })
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 4
  /* STYLE */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarSwipeItem',

  setup() {
    var translate = (0, _vue.ref)(0);
    var {
      swipe,
      bindSwipe,
      index
    } = (0, _provide.useSwipe)();
    var {
      size,
      vertical
    } = swipe;

    var setTranslate = x => {
      translate.value = x;
    };

    var swipeItemProvider = {
      index,
      setTranslate
    };
    bindSwipe(swipeItemProvider);
    return {
      size,
      vertical,
      translate
    };
  }

});

exports.default = _default;
