// @varlet/ui/lib/checkbox/Checkbox.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _icon = _interopRequireDefault(require("../icon"));

var _formDetails = _interopRequireDefault(require("../form-details"));

var _ripple = _interopRequireDefault(require("../ripple"));

var _vue = require("vue");

var _props = require("./props");

var _components = require("../utils/components");

var _provide = require("./provide");

var _provide2 = require("../form/provide");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-checkbox"
};

function render(_ctx, _cache) {
  var _component_var_icon = (0, _vue.resolveComponent)("var-icon");

  var _component_var_form_details = (0, _vue.resolveComponent)("var-form-details");

  var _directive_ripple = (0, _vue.resolveDirective)("ripple");

  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: "var-checkbox__wrap",
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [(0, _vue.createElementVNode)("div", _hoisted_1, [(0, _vue.withDirectives)((0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-checkbox__action", [_ctx.checked ? 'var-checkbox--checked' : 'var-checkbox--unchecked', _ctx.errorMessage || _ctx.checkboxGroupErrorMessage ? 'var-checkbox--error' : null, _ctx.formDisabled || _ctx.disabled ? 'var-checkbox--disabled' : null]]),
    style: (0, _vue.normalizeStyle)({
      color: _ctx.checked ? _ctx.checkedColor : _ctx.uncheckedColor
    })
  }, [_ctx.checked ? (0, _vue.renderSlot)(_ctx.$slots, "checked-icon", {
    key: 0
  }, () => [(0, _vue.createVNode)(_component_var_icon, {
    class: (0, _vue.normalizeClass)(["var-checkbox__icon", [_ctx.withAnimation ? 'var-checkbox--with-animation' : null]]),
    name: "checkbox-marked",
    size: _ctx.iconSize,
    "var-checkbox-cover": ""
  }, null, 8
  /* PROPS */
  , ["class", "size"])]) : (0, _vue.renderSlot)(_ctx.$slots, "unchecked-icon", {
    key: 1
  }, () => [(0, _vue.createVNode)(_component_var_icon, {
    class: (0, _vue.normalizeClass)(["var-checkbox__icon", [_ctx.withAnimation ? 'var-checkbox--with-animation' : null]]),
    name: "checkbox-blank-outline",
    size: _ctx.iconSize,
    "var-checkbox-cover": ""
  }, null, 8
  /* PROPS */
  , ["class", "size"])])], 6
  /* CLASS, STYLE */
  ), [[_directive_ripple, {
    disabled: _ctx.formReadonly || _ctx.readonly || _ctx.formDisabled || _ctx.disabled || !_ctx.ripple
  }]]), (0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-checkbox__text", [_ctx.errorMessage || _ctx.checkboxGroupErrorMessage ? 'var-checkbox--error' : null, _ctx.formDisabled || _ctx.disabled ? 'var-checkbox--disabled' : null]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 2
  /* CLASS */
  )]), (0, _vue.createVNode)(_component_var_form_details, {
    "error-message": _ctx.errorMessage
  }, null, 8
  /* PROPS */
  , ["error-message"])]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarCheckbox',
  directives: {
    Ripple: _ripple.default
  },
  components: {
    VarIcon: _icon.default,
    VarFormDetails: _formDetails.default
  },
  props: _props.props,

  setup(props) {
    var value = (0, _vue.ref)(false);
    var checked = (0, _vue.computed)(() => value.value === props.checkedValue);
    var checkedValue = (0, _vue.computed)(() => props.checkedValue);
    var withAnimation = (0, _vue.ref)(false);
    var {
      checkboxGroup,
      bindCheckboxGroup
    } = (0, _provide.useCheckboxGroup)();
    var {
      form,
      bindForm
    } = (0, _provide2.useForm)();
    var {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      // expose
      resetValidation
    } = (0, _components.useValidation)();

    var validateWithTrigger = trigger => {
      (0, _vue.nextTick)(() => {
        var {
          validateTrigger,
          rules,
          modelValue
        } = props;
        vt(validateTrigger, trigger, rules, modelValue);
      });
    };

    var change = changedValue => {
      var _props$onUpdateModel;

      value.value = changedValue;
      var {
        checkedValue,
        onChange
      } = props;
      (_props$onUpdateModel = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel.call(props, value.value);
      onChange == null ? void 0 : onChange(value.value);
      validateWithTrigger('onChange');
      changedValue === checkedValue ? checkboxGroup == null ? void 0 : checkboxGroup.onChecked(checkedValue) : checkboxGroup == null ? void 0 : checkboxGroup.onUnchecked(checkedValue);
    };

    var handleClick = e => {
      var {
        disabled,
        readonly,
        checkedValue,
        uncheckedValue,
        onClick
      } = props;

      if (form != null && form.disabled.value || disabled) {
        return;
      }

      onClick == null ? void 0 : onClick(e);

      if (form != null && form.readonly.value || readonly) {
        return;
      }

      withAnimation.value = true;
      var maximum = checkboxGroup ? checkboxGroup.checkedCount.value >= Number(checkboxGroup.max.value) : false;

      if (!checked.value && maximum) {
        return;
      }

      change(checked.value ? uncheckedValue : checkedValue);
    };

    var sync = values => {
      var {
        checkedValue,
        uncheckedValue
      } = props;
      value.value = values.includes(checkedValue) ? checkedValue : uncheckedValue;
    }; // expose


    var reset = () => {
      var _props$onUpdateModel2;

      (_props$onUpdateModel2 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel2.call(props, props.uncheckedValue);
      resetValidation();
    }; // expose


    var toggle = changedValue => {
      var {
        checkedValue,
        uncheckedValue
      } = props;
      var shouldReverse = ![checkedValue, uncheckedValue].includes(changedValue);

      if (shouldReverse) {
        changedValue = checked.value ? uncheckedValue : checkedValue;
      }

      change(changedValue);
    }; // expose


    var validate = () => v(props.rules, props.modelValue);

    (0, _vue.watch)(() => props.modelValue, newValue => {
      value.value = newValue;
    }, {
      immediate: true
    });
    var checkboxProvider = {
      checkedValue,
      checked,
      sync,
      validate,
      resetValidation,
      reset
    };
    bindCheckboxGroup == null ? void 0 : bindCheckboxGroup(checkboxProvider);
    bindForm == null ? void 0 : bindForm(checkboxProvider);
    return {
      withAnimation,
      checked,
      errorMessage,
      checkboxGroupErrorMessage: checkboxGroup == null ? void 0 : checkboxGroup.errorMessage,
      formDisabled: form == null ? void 0 : form.disabled,
      formReadonly: form == null ? void 0 : form.readonly,
      handleClick,
      toggle,
      reset,
      validate,
      resetValidation
    };
  }

});

exports.default = _default;
