// @varlet/ui/lib/index-bar/IndexBar.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _shared = require("../utils/shared");

var _elements = require("../utils/elements");

var _provide = require("./provide");

var _props = require("./props");

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-index-bar",
  ref: "barEl"
};
var _hoisted_2 = ["onClick"];

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [(0, _vue.renderSlot)(_ctx.$slots, "default"), (0, _vue.createElementVNode)("ul", {
    class: "var-index-bar__anchor-list",
    style: (0, _vue.normalizeStyle)({
      zIndex: _ctx.toNumber(_ctx.zIndex) + 2,
      display: _ctx.hideList ? 'none' : 'block'
    })
  }, [((0, _vue.openBlock)(true), (0, _vue.createElementBlock)(_vue.Fragment, null, (0, _vue.renderList)(_ctx.anchorNameList, anchorName => {
    return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("li", {
      key: anchorName,
      class: (0, _vue.normalizeClass)(["var-index-bar__anchor-item", {
        'var-index-bar__anchor-item--active': _ctx.active === anchorName
      }]),
      style: (0, _vue.normalizeStyle)({
        color: _ctx.active === anchorName && _ctx.highlightColor ? _ctx.highlightColor : ''
      }),
      onClick: $event => _ctx.anchorClick(anchorName)
    }, (0, _vue.toDisplayString)(anchorName), 15
    /* TEXT, CLASS, STYLE, PROPS */
    , _hoisted_2);
  }), 128
  /* KEYED_FRAGMENT */
  ))], 4
  /* STYLE */
  )], 512
  /* NEED_PATCH */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarIndexBar',
  props: _props.props,

  setup(props) {
    var {
      length,
      indexAnchors,
      bindIndexAnchors
    } = (0, _provide.useIndexAnchors)();
    var scrollEl = (0, _vue.ref)(null);
    var clickedName = (0, _vue.ref)('');
    var scroller = (0, _vue.ref)(null);
    var barEl = (0, _vue.ref)(null);
    var anchorNameList = (0, _vue.ref)([]);
    var active = (0, _vue.ref)();
    var sticky = (0, _vue.computed)(() => props.sticky);
    var cssMode = (0, _vue.computed)(() => props.cssMode);
    var stickyOffsetTop = (0, _vue.computed)(() => props.stickyOffsetTop);
    var zIndex = (0, _vue.computed)(() => props.zIndex);
    var indexBarProvider = {
      active,
      sticky,
      cssMode,
      stickyOffsetTop,
      zIndex
    };
    bindIndexAnchors(indexBarProvider);

    var emitEvent = anchor => {
      var anchorName = (0, _shared.isPlainObject)(anchor) ? anchor.name.value : anchor;
      if (anchorName === active.value) return;
      active.value = anchorName;
      props.onChange == null ? void 0 : props.onChange(anchorName);
    };

    var handleScroll = () => {
      var {
        scrollTop,
        scrollHeight
      } = scrollEl.value;
      var {
        offsetTop
      } = barEl.value;
      indexAnchors.forEach((anchor, index) => {
        var anchorTop = anchor.ownTop.value;
        var top = scrollTop - anchorTop + stickyOffsetTop.value - offsetTop;
        var distance = index === indexAnchors.length - 1 ? scrollHeight : indexAnchors[index + 1].ownTop.value - anchor.ownTop.value;

        if (top >= 0 && top < distance && !clickedName.value) {
          if (index && !props.cssMode) {
            indexAnchors[index - 1].setDisabled(true);
          }

          anchor.setDisabled(false);
          emitEvent(anchor);
        }
      });
    };

    var anchorClick = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(function* (anchorName, manualCall) {
        if (manualCall) props.onClick == null ? void 0 : props.onClick(anchorName);
        if (anchorName === active.value) return;
        var indexAnchor = indexAnchors.find(({
          name
        }) => anchorName === name.value);
        if (!indexAnchor) return;
        var top = indexAnchor.ownTop.value;
        var left = (0, _elements.getScrollLeft)(scrollEl.value);
        clickedName.value = anchorName;
        emitEvent(anchorName);
        yield (0, _elements.scrollTo)(scrollEl.value, {
          left,
          top,
          animation: _shared.easeInOutCubic,
          duration: (0, _shared.toNumber)(props.duration)
        });
        (0, _elements.nextTickFrame)(() => {
          clickedName.value = '';
        });
      });

      return function anchorClick(_x, _x2) {
        return _ref.apply(this, arguments);
      };
    }(); // expose


    var scrollTo = index => {
      (0, _elements.requestAnimationFrame)(() => anchorClick(index, true));
    };

    (0, _vue.watch)(() => length.value, () => (0, _vue.nextTick)(() => {
      indexAnchors.forEach(({
        name,
        setOwnTop
      }) => {
        if (name.value) anchorNameList.value.push(name.value);
        setOwnTop();
      });
    }));
    (0, _vue.onMounted)(() => {
      var _scroller$value;

      scroller.value = (0, _elements.getParentScroller)(barEl.value);
      scrollEl.value = scroller.value === window ? scroller.value.document.documentElement : scroller.value;
      (_scroller$value = scroller.value) == null ? void 0 : _scroller$value.addEventListener('scroll', handleScroll);
    });
    (0, _vue.onBeforeUnmount)(() => {
      var _scroller$value2;

      (_scroller$value2 = scroller.value) == null ? void 0 : _scroller$value2.removeEventListener('scroll', handleScroll);
    });
    return {
      barEl,
      active,
      zIndex,
      anchorNameList,
      toNumber: _shared.toNumber,
      scrollTo,
      anchorClick
    };
  }

});

exports.default = _default;
