// @varlet/ui/lib/table/Table.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _elements = require("../utils/elements");

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-table var-elevation--1 var--box"
};
var _hoisted_2 = {
  class: "var-table__main"
};
var _hoisted_3 = {
  key: 0,
  class: "var-table__footer"
};

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [(0, _vue.createElementVNode)("div", _hoisted_2, [(0, _vue.createElementVNode)("table", {
    class: "var-table__table",
    style: (0, _vue.normalizeStyle)({
      width: _ctx.toSizeUnit(_ctx.fullWidth)
    })
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 4
  /* STYLE */
  )]), _ctx.$slots.footer ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_3, [(0, _vue.renderSlot)(_ctx.$slots, "footer")])) : (0, _vue.createCommentVNode)("v-if", true)]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarTable',
  props: {
    fullWidth: {
      type: [Number, String],
      default: '100%'
    }
  },

  setup() {
    return {
      toSizeUnit: _elements.toSizeUnit
    };
  }

});

exports.default = _default;
