// @varlet/ui/lib/sticky/Sticky.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _props = require("./props");

var _elements = require("../utils/elements");

var _shared = require("../utils/shared");

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-sticky", [_ctx.enableCSSMode ? 'var-sticky--css-mode' : null]]),
    ref: "stickyEl",
    style: (0, _vue.normalizeStyle)({
      zIndex: _ctx.toNumber(_ctx.zIndex),
      top: _ctx.enableCSSMode ? _ctx.offsetTop + "px" : undefined,
      width: _ctx.enableFixedMode ? _ctx.fixedWidth : undefined,
      height: _ctx.enableFixedMode ? _ctx.fixedHeight : undefined
    })
  }, [(0, _vue.createElementVNode)("div", {
    class: "var-sticky__wrapper",
    ref: "wrapperEl",
    style: (0, _vue.normalizeStyle)({
      zIndex: _ctx.toNumber(_ctx.zIndex),
      position: _ctx.enableFixedMode ? 'fixed' : undefined,
      width: _ctx.enableFixedMode ? _ctx.fixedWrapperWidth : undefined,
      height: _ctx.enableFixedMode ? _ctx.fixedWrapperHeight : undefined,
      left: _ctx.enableFixedMode ? _ctx.fixedLeft : undefined,
      top: _ctx.enableFixedMode ? _ctx.fixedTop : undefined
    })
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 4
  /* STYLE */
  )], 6
  /* CLASS, STYLE */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarSticky',
  props: _props.props,

  setup(props) {
    var stickyEl = (0, _vue.ref)(null);
    var wrapperEl = (0, _vue.ref)(null);
    var isFixed = (0, _vue.ref)(false);
    var fixedTop = (0, _vue.ref)('0px');
    var fixedLeft = (0, _vue.ref)('0px');
    var fixedWidth = (0, _vue.ref)('auto');
    var fixedHeight = (0, _vue.ref)('auto');
    var fixedWrapperWidth = (0, _vue.ref)('auto');
    var fixedWrapperHeight = (0, _vue.ref)('auto');
    var enableCSSMode = (0, _vue.computed)(() => !props.disabled && props.cssMode);
    var enableFixedMode = (0, _vue.computed)(() => !props.disabled && isFixed.value);
    var offsetTop = (0, _vue.computed)(() => (0, _elements.toPxNum)(props.offsetTop));
    var scroller;

    var handleScroll = () => {
      var {
        onScroll,
        cssMode,
        disabled
      } = props;

      if (disabled) {
        return;
      }

      var scrollerTop = 0;

      if (scroller !== window) {
        var {
          top
        } = scroller.getBoundingClientRect();
        scrollerTop = top;
      }

      var wrapper = wrapperEl.value;
      var sticky = stickyEl.value;
      var {
        top: stickyTop,
        left: stickyLeft
      } = sticky.getBoundingClientRect();
      var currentOffsetTop = stickyTop - scrollerTop;

      if (currentOffsetTop <= offsetTop.value) {
        if (!cssMode) {
          fixedWidth.value = sticky.offsetWidth + "px";
          fixedHeight.value = sticky.offsetHeight + "px";
          fixedTop.value = scrollerTop + offsetTop.value + "px";
          fixedLeft.value = stickyLeft + "px";
          fixedWrapperWidth.value = wrapper.offsetWidth + "px";
          fixedWrapperHeight.value = wrapper.offsetHeight + "px";
          isFixed.value = true;
        }

        onScroll == null ? void 0 : onScroll(offsetTop.value, true);
      } else {
        isFixed.value = false;
        onScroll == null ? void 0 : onScroll(currentOffsetTop, false);
      }
    };

    var addScrollListener = () => {
      scroller = (0, _elements.getParentScroller)(stickyEl.value);
      scroller !== window && scroller.addEventListener('scroll', handleScroll);
      window.addEventListener('scroll', handleScroll);
      handleScroll();
    };

    var removeScrollListener = () => {
      scroller !== window && scroller.removeEventListener('scroll', handleScroll);
      window.removeEventListener('scroll', handleScroll);
    };

    (0, _vue.watch)(() => props.disabled, handleScroll);
    (0, _vue.onActivated)(addScrollListener);
    (0, _vue.onDeactivated)(removeScrollListener);
    (0, _vue.onMounted)(addScrollListener);
    (0, _vue.onUnmounted)(removeScrollListener);
    return {
      stickyEl,
      wrapperEl,
      isFixed,
      offsetTop,
      fixedTop,
      fixedLeft,
      fixedWidth,
      fixedHeight,
      fixedWrapperWidth,
      fixedWrapperHeight,
      enableCSSMode,
      enableFixedMode,
      toNumber: _shared.toNumber
    };
  }

});

exports.default = _default;
