// @varlet/ui/lib/progress/Progress.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _props = require("./props");

var _shared = require("../utils/shared");

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-progress"
};
var _hoisted_2 = {
  key: 0,
  class: "var-progress-linear"
};
var _hoisted_3 = ["viewBox"];
var _hoisted_4 = ["cx", "cy", "r", "stroke-width"];
var _hoisted_5 = ["cx", "cy", "r", "stroke-width"];

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [_ctx.mode === 'linear' ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_2, [(0, _vue.createElementVNode)("div", (0, _vue.mergeProps)({
    class: "var-progress-linear__block",
    style: {
      height: _ctx.lineWidth + "px"
    }
  }, _ctx.$attrs), [_ctx.track ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: "var-progress-linear__background",
    style: (0, _vue.normalizeStyle)({
      background: _ctx.trackColor
    })
  }, null, 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)("var-progress-linear__certain" + (_ctx.ripple ? ' var-progress-linear__ripple' : '')),
    style: (0, _vue.normalizeStyle)({
      background: _ctx.color,
      width: _ctx.linearProps.width
    })
  }, null, 6
  /* CLASS, STYLE */
  )], 16
  /* FULL_PROPS */
  ), _ctx.label ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", (0, _vue.mergeProps)({
    key: 0,
    class: "var-progress-linear__label"
  }, _ctx.$attrs), [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.linearProps.roundValue), 1
  /* TEXT */
  )])], 16
  /* FULL_PROPS */
  )) : (0, _vue.createCommentVNode)("v-if", true)])) : (0, _vue.createCommentVNode)("v-if", true), _ctx.mode === 'circle' ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 1,
    class: "var-progress-circle",
    style: (0, _vue.normalizeStyle)({
      width: _ctx.size + "px",
      height: _ctx.size + "px"
    })
  }, [((0, _vue.openBlock)(), (0, _vue.createElementBlock)("svg", {
    class: "var-progress-circle__svg",
    style: (0, _vue.normalizeStyle)({
      transform: "rotate(" + (_ctx.rotate - 90) + "deg)"
    }),
    viewBox: _ctx.circleProps.viewBox
  }, [_ctx.track ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("circle", {
    key: 0,
    class: "var-progress-circle__background",
    cx: _ctx.size / 2,
    cy: _ctx.size / 2,
    r: _ctx.circleProps.radius,
    fill: "transparent",
    "stroke-width": _ctx.lineWidth,
    style: (0, _vue.normalizeStyle)({
      strokeDasharray: _ctx.circleProps.perimeter,
      stroke: _ctx.trackColor
    })
  }, null, 12
  /* STYLE, PROPS */
  , _hoisted_4)) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("circle", {
    class: "var-progress-circle__certain",
    cx: _ctx.size / 2,
    cy: _ctx.size / 2,
    r: _ctx.circleProps.radius,
    fill: "transparent",
    "stroke-width": _ctx.lineWidth,
    style: (0, _vue.normalizeStyle)({
      strokeDasharray: _ctx.circleProps.strokeDasharray,
      stroke: _ctx.color
    })
  }, null, 12
  /* STYLE, PROPS */
  , _hoisted_5)], 12
  /* STYLE, PROPS */
  , _hoisted_3)), _ctx.label ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", (0, _vue.mergeProps)({
    key: 0,
    class: "var-progress-circle__label"
  }, _ctx.$attrs), [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.circleProps.roundValue), 1
  /* TEXT */
  )])], 16
  /* FULL_PROPS */
  )) : (0, _vue.createCommentVNode)("v-if", true)], 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true)]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarProgress',
  inheritAttrs: false,
  props: _props.props,

  setup(props) {
    var linearProps = (0, _vue.computed)(() => {
      var value = (0, _shared.toNumber)(props.value);
      var width = value > 100 ? 100 : value;
      var roundValue = value > 100 ? 100 : Math.round(value);
      return {
        width: width + "%",
        roundValue: roundValue + "%"
      };
    });
    var circleProps = (0, _vue.computed)(() => {
      var {
        size,
        lineWidth,
        value
      } = props;
      var viewBox = "0 0 " + size + " " + size;
      var roundValue = (0, _shared.toNumber)(value) > 100 ? 100 : Math.round((0, _shared.toNumber)(value));
      var radius = (size - (0, _shared.toNumber)(lineWidth)) / 2;
      var perimeter = 2 * Math.PI * radius;
      var strokeDasharray = roundValue / 100 * perimeter + ", " + perimeter;
      return {
        viewBox,
        radius,
        strokeDasharray,
        perimeter,
        roundValue: roundValue + "%"
      };
    });
    return {
      linearProps,
      circleProps
    };
  }

});

exports.default = _default;
