// @varlet/ui/es/option/Option.js
import VarCheckbox from '../checkbox';
import Ripple from '../ripple';
import { defineComponent, computed, ref, watch } from 'vue';
import { useSelect } from './provide';
import { props } from './props';
import { normalizeClass as _normalizeClass, normalizeStyle as _normalizeStyle, createElementVNode as _createElementVNode, resolveComponent as _resolveComponent, withModifiers as _withModifiers, openBlock as _openBlock, createBlock as _createBlock, createCommentVNode as _createCommentVNode, renderSlot as _renderSlot, toDisplayString as _toDisplayString, createTextVNode as _createTextVNode, resolveDirective as _resolveDirective, withDirectives as _withDirectives, createElementBlock as _createElementBlock, pushScopeId as _pushScopeId, popScopeId as _popScopeId } from "vue";

var _withScopeId = n => (_pushScopeId(""), n = n(), _popScopeId(), n);

var _hoisted_1 = {
  class: "var-option__text"
};
export function render(_ctx, _cache) {
  var _component_var_checkbox = _resolveComponent("var-checkbox");

  var _directive_ripple = _resolveDirective("ripple");

  return _withDirectives((_openBlock(), _createElementBlock("div", {
    class: _normalizeClass(["var-option var--box", [_ctx.optionSelected ? 'var-option--selected-color' : null]]),
    style: _normalizeStyle({
      width: _ctx.wrapWidth,
      color: _ctx.optionSelected ? _ctx.focusColor : undefined
    }),
    onClick: _cache[2] || (_cache[2] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [_createElementVNode("div", {
    class: _normalizeClass(["var-option__cover", [_ctx.optionSelected ? 'var-option--selected-background' : null]]),
    style: _normalizeStyle({
      background: _ctx.optionSelected ? _ctx.focusColor : undefined
    })
  }, null, 6
  /* CLASS, STYLE */
  ), _ctx.multiple ? (_openBlock(), _createBlock(_component_var_checkbox, {
    key: 0,
    ref: "checkbox",
    "checked-color": _ctx.focusColor,
    modelValue: _ctx.optionSelected,
    "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => _ctx.optionSelected = $event),
    onClick: _cache[1] || (_cache[1] = _withModifiers(() => {}, ["stop"])),
    onChange: _ctx.handleSelect
  }, null, 8
  /* PROPS */
  , ["checked-color", "modelValue", "onChange"])) : _createCommentVNode("v-if", true), _createElementVNode("div", _hoisted_1, [_renderSlot(_ctx.$slots, "default", {}, () => [_createTextVNode(_toDisplayString(_ctx.label), 1
  /* TEXT */
  )])])], 6
  /* CLASS, STYLE */
  )), [[_directive_ripple]]);
}
export default defineComponent({
  render,
  name: 'VarOption',
  directives: {
    Ripple
  },
  components: {
    VarCheckbox
  },
  props,

  setup(props) {
    var optionSelected = ref(false);
    var selected = computed(() => optionSelected.value);
    var label = computed(() => props.label);
    var value = computed(() => props.value);
    var {
      select,
      bindSelect
    } = useSelect();
    var {
      wrapWidth,
      multiple,
      focusColor,
      onSelect
    } = select;

    var handleClick = () => {
      optionSelected.value = !optionSelected.value;
      onSelect(optionProvider);
    };

    var handleSelect = () => onSelect(optionProvider);

    var sync = checked => {
      optionSelected.value = checked;
    };

    var optionProvider = {
      label,
      value,
      selected,
      sync
    };
    watch([() => props.label, () => props.value], () => {
      if (props.label == null && props.value == null) {
        throw Error("Props label and value can't both be undefined\n");
      }
    }, {
      immediate: true
    });
    bindSelect(optionProvider);
    return {
      optionSelected,
      wrapWidth,
      multiple,
      focusColor,
      handleClick,
      handleSelect
    };
  }

});
