// @varlet/ui/lib/tabs-items/TabsItems.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _swipe = _interopRequireDefault(require("../swipe"));

var _vue = require("vue");

var _provide = require("./provide");

var _props = require("./props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function render(_ctx, _cache) {
  var _component_var_swipe = (0, _vue.resolveComponent)("var-swipe");

  return (0, _vue.openBlock)(), (0, _vue.createBlock)(_component_var_swipe, {
    class: "var-tabs-items",
    ref: "swipe",
    loop: _ctx.loop,
    touchable: _ctx.canSwipe,
    indicator: false,
    onChange: _ctx.handleSwipeChange
  }, {
    default: (0, _vue.withCtx)(() => [(0, _vue.renderSlot)(_ctx.$slots, "default")]),
    _: 3
    /* FORWARDED */

  }, 8
  /* PROPS */
  , ["loop", "touchable", "onChange"]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarTabsItems',
  components: {
    VarSwipe: _swipe.default
  },
  props: _props.props,

  setup(props) {
    var swipe = (0, _vue.ref)(null);
    var {
      tabItemList,
      bindTabItem,
      length
    } = (0, _provide.useTabItem)();

    var matchName = active => {
      return tabItemList.find(({
        name
      }) => active === name.value);
    };

    var matchIndex = active => {
      return tabItemList.find(({
        index
      }) => active === index.value);
    };

    var matchActive = active => {
      return matchName(active) || matchIndex(active);
    };

    var handleActiveChange = newValue => {
      var _swipe$value;

      var newActiveTabItemProvider = matchActive(newValue);

      if (!newActiveTabItemProvider) {
        return;
      }

      tabItemList.forEach(({
        setCurrent
      }) => setCurrent(false));
      newActiveTabItemProvider.setCurrent(true);
      (_swipe$value = swipe.value) == null ? void 0 : _swipe$value.to(newActiveTabItemProvider.index.value);
    };

    var handleSwipeChange = currentIndex => {
      var _tabItem$name$value, _props$onUpdateActiv;

      var tabItem = tabItemList.find(({
        index
      }) => index.value === currentIndex);
      var active = (_tabItem$name$value = tabItem.name.value) != null ? _tabItem$name$value : tabItem.index.value;
      (_props$onUpdateActiv = props['onUpdate:active']) == null ? void 0 : _props$onUpdateActiv.call(props, active);
    };

    var tabsItemsProvider = {};
    bindTabItem(tabsItemsProvider);
    (0, _vue.watch)(() => props.active, handleActiveChange);
    (0, _vue.watch)(() => length.value, () => handleActiveChange(props.active));
    return {
      swipe,
      handleSwipeChange
    };
  }

});

exports.default = _default;
