// @varlet/ui/es/swipe-item/SwipeItem.js
import { defineComponent, ref } from 'vue';
import { useSwipe } from './provide';
import { renderSlot as _renderSlot, normalizeStyle as _normalizeStyle, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue";
export function render(_ctx, _cache) {
  return _openBlock(), _createElementBlock("div", {
    class: "var-swipe-item",
    style: _normalizeStyle({
      width: !_ctx.vertical ? _ctx.size + "px" : undefined,
      height: _ctx.vertical ? _ctx.size + "px" : undefined,
      transform: "translate" + (_ctx.vertical ? 'Y' : 'X') + "(" + _ctx.translate + "px)"
    })
  }, [_renderSlot(_ctx.$slots, "default")], 4
  /* STYLE */
  );
}
export default defineComponent({
  render,
  name: 'VarSwipeItem',

  setup() {
    var translate = ref(0);
    var {
      swipe,
      bindSwipe,
      index
    } = useSwipe();
    var {
      size,
      vertical
    } = swipe;

    var setTranslate = x => {
      translate.value = x;
    };

    var swipeItemProvider = {
      index,
      setTranslate
    };
    bindSwipe(swipeItemProvider);
    return {
      size,
      vertical,
      translate
    };
  }

});
