// @varlet/ui/lib/cell/Cell.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _props = require("./props");

var _icon = _interopRequireDefault(require("../icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-cell__content"
};

function render(_ctx, _cache) {
  var _component_var_icon = (0, _vue.resolveComponent)("var-icon");

  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-cell", [_ctx.border ? 'var-cell--border' : null]])
  }, [_ctx.$slots.icon || _ctx.icon ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: (0, _vue.normalizeClass)(["var-cell__icon", [_ctx.iconClass ? _ctx.iconClass : null]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "icon", {}, () => [(0, _vue.createVNode)(_component_var_icon, {
    class: "var--flex",
    name: _ctx.icon
  }, null, 8
  /* PROPS */
  , ["name"])])], 2
  /* CLASS */
  )) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("div", _hoisted_1, [(0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-cell__title", [_ctx.titleClass ? _ctx.titleClass : null]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.title), 1
  /* TEXT */
  )])], 2
  /* CLASS */
  ), _ctx.$slots.desc || _ctx.desc ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: (0, _vue.normalizeClass)(["var-cell__desc", [_ctx.descClass ? _ctx.descClass : null]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "desc", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.desc), 1
  /* TEXT */
  )])], 2
  /* CLASS */
  )) : (0, _vue.createCommentVNode)("v-if", true)]), _ctx.$slots.extra ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 1,
    class: (0, _vue.normalizeClass)(["var-cell__extra", [_ctx.extraClass ? _ctx.extraClass : null]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "extra")], 2
  /* CLASS */
  )) : (0, _vue.createCommentVNode)("v-if", true)], 2
  /* CLASS */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarCell',
  components: {
    VarIcon: _icon.default
  },
  props: _props.props
});

exports.default = _default;
