// @varlet/ui/lib/radio-group/RadioGroup.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _formDetails = _interopRequireDefault(require("../form-details"));

var _vue = require("vue");

var _props = require("./props");

var _components = require("../utils/components");

var _provide = require("./provide");

var _provide2 = require("../form/provide");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-radio-group__wrap"
};

function render(_ctx, _cache) {
  var _component_var_form_details = (0, _vue.resolveComponent)("var-form-details");

  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [(0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-radio-group", ["var-radio-group--" + _ctx.direction]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 2
  /* CLASS */
  ), (0, _vue.createVNode)(_component_var_form_details, {
    "error-message": _ctx.errorMessage
  }, null, 8
  /* PROPS */
  , ["error-message"])]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarRadioGroup',
  components: {
    VarFormDetails: _formDetails.default
  },
  props: _props.props,

  setup(props) {
    var {
      length,
      radios,
      bindRadios
    } = (0, _provide.useRadios)();
    var {
      bindForm
    } = (0, _provide2.useForm)();
    var {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      // expose
      resetValidation
    } = (0, _components.useValidation)();
    var radioGroupErrorMessage = (0, _vue.computed)(() => errorMessage.value);

    var validateWithTrigger = trigger => {
      (0, _vue.nextTick)(() => {
        var {
          validateTrigger,
          rules,
          modelValue
        } = props;
        vt(validateTrigger, trigger, rules, modelValue);
      });
    };

    var syncRadios = () => radios.forEach(({
      sync
    }) => sync(props.modelValue));

    var onToggle = changedValue => {
      var _props$onUpdateModel;

      (_props$onUpdateModel = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel.call(props, changedValue);
      props.onChange == null ? void 0 : props.onChange(changedValue);
      validateWithTrigger('onChange');
    }; // expose


    var validate = () => v(props.rules, props.modelValue); // expose


    var reset = () => {
      var _props$onUpdateModel2;

      (_props$onUpdateModel2 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel2.call(props, undefined);
      resetValidation();
    };

    (0, _vue.watch)(() => props.modelValue, syncRadios);
    (0, _vue.watch)(() => length.value, syncRadios);
    var radioGroupProvider = {
      onToggle,
      validate,
      reset,
      resetValidation,
      errorMessage: radioGroupErrorMessage
    };
    bindForm == null ? void 0 : bindForm(radioGroupProvider);
    bindRadios(radioGroupProvider);
    return {
      errorMessage,
      reset,
      validate,
      resetValidation
    };
  }

});

exports.default = _default;
