// @varlet/ui/lib/card/Card.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _ripple = _interopRequireDefault(require("../ripple"));

var _vue = require("vue");

var _props = require("./props");

var _elements = require("../utils/elements");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = ["src", "alt"];
var _hoisted_2 = {
  key: 0,
  class: "var-card__title"
};
var _hoisted_3 = {
  key: 0,
  class: "var-card__subtitle"
};
var _hoisted_4 = {
  key: 0,
  class: "var-card__description"
};
var _hoisted_5 = {
  key: 0,
  class: "var-card__footer"
};

function render(_ctx, _cache) {
  var _directive_ripple = (0, _vue.resolveDirective)("ripple");

  return (0, _vue.withDirectives)(((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-card", [_ctx.elevation ? "var-elevation--" + _ctx.elevation : 'var-elevation--2']]),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.onClick && _ctx.onClick(...args))
  }, [(0, _vue.renderSlot)(_ctx.$slots, "image", {}, () => [_ctx.src ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("img", {
    key: 0,
    class: "var-card__image",
    style: (0, _vue.normalizeStyle)({
      objectFit: _ctx.fit,
      height: _ctx.toSizeUnit(_ctx.height)
    }),
    src: _ctx.src,
    alt: _ctx.alt
  }, null, 12
  /* STYLE, PROPS */
  , _hoisted_1)) : (0, _vue.createCommentVNode)("v-if", true)]), (0, _vue.renderSlot)(_ctx.$slots, "title", {}, () => [_ctx.title ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_2, (0, _vue.toDisplayString)(_ctx.title), 1
  /* TEXT */
  )) : (0, _vue.createCommentVNode)("v-if", true)]), (0, _vue.renderSlot)(_ctx.$slots, "subtitle", {}, () => [_ctx.subtitle ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_3, (0, _vue.toDisplayString)(_ctx.subtitle), 1
  /* TEXT */
  )) : (0, _vue.createCommentVNode)("v-if", true)]), (0, _vue.renderSlot)(_ctx.$slots, "description", {}, () => [_ctx.description ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_4, (0, _vue.toDisplayString)(_ctx.description), 1
  /* TEXT */
  )) : (0, _vue.createCommentVNode)("v-if", true)]), _ctx.$slots.extra ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_5, [(0, _vue.renderSlot)(_ctx.$slots, "extra")])) : (0, _vue.createCommentVNode)("v-if", true)], 2
  /* CLASS */
  )), [[_directive_ripple, {
    disabled: !_ctx.ripple
  }]]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarCard',
  directives: {
    Ripple: _ripple.default
  },
  props: _props.props,

  setup() {
    return {
      toSizeUnit: _elements.toSizeUnit
    };
  }

});

exports.default = _default;
