// @varlet/ui/es/switch/Switch.js
import { defineComponent, computed, nextTick } from 'vue';
import { useValidation } from '../utils/components';
import { toNumber } from '../utils/shared';
import { useForm } from '../form/provide';
import { props } from './props';
import VarFormDetails from '../form-details';
import VarLoading from '../loading';
import Ripple from '../ripple';
import { normalizeClass as _normalizeClass, normalizeStyle as _normalizeStyle, createElementVNode as _createElementVNode, resolveComponent as _resolveComponent, openBlock as _openBlock, createBlock as _createBlock, createCommentVNode as _createCommentVNode, resolveDirective as _resolveDirective, withDirectives as _withDirectives, createVNode as _createVNode, createElementBlock as _createElementBlock, pushScopeId as _pushScopeId, popScopeId as _popScopeId } from "vue";

var _withScopeId = n => (_pushScopeId(""), n = n(), _popScopeId(), n);

var _hoisted_1 = {
  class: "var-switch"
};
export function render(_ctx, _cache) {
  var _component_var_loading = _resolveComponent("var-loading");

  var _component_var_form_details = _resolveComponent("var-form-details");

  var _directive_ripple = _resolveDirective("ripple");

  return _openBlock(), _createElementBlock("div", _hoisted_1, [_createElementVNode("div", {
    class: _normalizeClass(["var-switch-block", [_ctx.disabled || _ctx.formDisabled ? 'var-switch__disable' : null]]),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.switchActive && _ctx.switchActive(...args)),
    style: _normalizeStyle(_ctx.styleComputed.switch)
  }, [_createElementVNode("div", {
    style: _normalizeStyle(_ctx.styleComputed.track),
    class: _normalizeClass(["var-switch__track", [_ctx.modelValue === _ctx.activeValue ? 'var-switch__track-active' : null, _ctx.errorMessage ? 'var-switch__track-error' : null]])
  }, null, 6
  /* CLASS, STYLE */
  ), _withDirectives(_createElementVNode("div", {
    class: "var-switch__ripple",
    style: _normalizeStyle(_ctx.styleComputed.ripple)
  }, [_createElementVNode("div", {
    style: _normalizeStyle(_ctx.styleComputed.handle),
    class: _normalizeClass(["var-switch__handle var-elevation--2", [_ctx.modelValue === _ctx.activeValue ? 'var-switch__handle-active' : null, _ctx.errorMessage ? 'var-switch__handle-error' : null]])
  }, [_ctx.loading ? (_openBlock(), _createBlock(_component_var_loading, {
    key: 0,
    radius: _ctx.toNumber(_ctx.size) / 2 - 2
  }, null, 8
  /* PROPS */
  , ["radius"])) : _createCommentVNode("v-if", true)], 6
  /* CLASS, STYLE */
  )], 4
  /* STYLE */
  ), [[_directive_ripple, {
    disabled: !_ctx.ripple || _ctx.disabled || _ctx.loading || _ctx.formDisabled
  }]])], 6
  /* CLASS, STYLE */
  ), _createVNode(_component_var_form_details, {
    "error-message": _ctx.errorMessage
  }, null, 8
  /* PROPS */
  , ["error-message"])]);
}
export default defineComponent({
  render,
  name: 'VarSwitch',
  components: {
    VarLoading,
    VarFormDetails
  },
  directives: {
    Ripple
  },
  props,

  setup(props) {
    var {
      bindForm,
      form
    } = useForm();
    var {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      resetValidation
    } = useValidation();

    var validate = () => v(props.rules, props.modelValue);

    var validateWithTrigger = () => nextTick(() => vt(['onChange'], 'onChange', props.rules, props.modelValue));

    var styleComputed = computed(() => {
      var {
        size,
        modelValue,
        color,
        closeColor,
        loadingColor,
        activeValue
      } = props;
      var sizeNum = toNumber(size);
      var switchWidth = sizeNum * 2;
      var switchHeight = sizeNum * 1.2;
      return {
        handle: {
          width: size + "px",
          height: size + "px",
          backgroundColor: modelValue === activeValue ? color || '' : closeColor || '',
          color: loadingColor && loadingColor
        },
        ripple: {
          left: modelValue === activeValue ? sizeNum / 2 + "px" : "-" + sizeNum / 2 + "px",
          color: modelValue === activeValue ? color || '' : closeColor || '#999',
          width: sizeNum * 2 + "px",
          height: sizeNum * 2 + "px"
        },
        track: {
          height: switchHeight * 0.6 + "px",
          width: switchWidth - 2 + "px",
          borderRadius: switchWidth / 3 + "px",
          filter: modelValue === activeValue || errorMessage != null && errorMessage.value ? 'opacity(.6)' : 'brightness(.6)',
          backgroundColor: modelValue === activeValue ? color || '' : closeColor || ''
        },
        switch: {
          height: switchHeight + "px",
          width: switchWidth + "px"
        }
      };
    });

    var switchActive = () => {
      var {
        onClick,
        onChange,
        disabled,
        loading,
        readonly,
        modelValue,
        activeValue,
        inactiveValue,
        'onUpdate:modelValue': updateModelValue
      } = props;
      onClick == null ? void 0 : onClick();
      if (disabled || loading || readonly || form != null && form.disabled.value || form != null && form.readonly.value) return;
      var newValue = modelValue === activeValue ? inactiveValue : activeValue;
      onChange == null ? void 0 : onChange(newValue);
      updateModelValue == null ? void 0 : updateModelValue(newValue);
      validateWithTrigger();
    };

    var reset = () => {
      var _props$onUpdateModel;

      (_props$onUpdateModel = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel.call(props, props.inactiveValue);
      resetValidation();
    };

    var switchProvider = {
      reset,
      validate,
      resetValidation
    };
    bindForm == null ? void 0 : bindForm(switchProvider);
    return {
      switchActive,
      toNumber,
      styleComputed,
      errorMessage,
      formDisabled: form == null ? void 0 : form.disabled,
      formReadonly: form == null ? void 0 : form.readonly
    };
  }

});
