// @varlet/ui/es/uploader/Uploader.js
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

import VarFormDetails from '../form-details';
import VarIcon from '../icon';
import VarPopup from '../popup';
import ImagePreview from '../image-preview';
import Ripple from '../ripple';
import { defineComponent, nextTick, reactive, computed, watch, ref } from 'vue';
import { props } from './props';
import { isNumber, isHTMLSupportImage, isHTMLSupportVideo, toNumber, isString } from '../utils/shared';
import { useValidation } from '../utils/components';
import { useForm } from '../form/provide';
var fid = 0;
import { renderList as _renderList, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock, toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, resolveComponent as _resolveComponent, createVNode as _createVNode, withModifiers as _withModifiers, createCommentVNode as _createCommentVNode, normalizeStyle as _normalizeStyle, normalizeClass as _normalizeClass, resolveDirective as _resolveDirective, withDirectives as _withDirectives, renderSlot as _renderSlot, withCtx as _withCtx, pushScopeId as _pushScopeId, popScopeId as _popScopeId } from "vue";

var _withScopeId = n => (_pushScopeId(""), n = n(), _popScopeId(), n);

var _hoisted_1 = {
  class: "var-uploader var--box"
};
var _hoisted_2 = {
  class: "var-uploader__file-list"
};
var _hoisted_3 = ["onClick"];
var _hoisted_4 = {
  class: "var-uploader__file-name"
};
var _hoisted_5 = ["onClick"];
var _hoisted_6 = ["src", "alt"];
var _hoisted_7 = ["multiple", "accept", "capture", "disabled"];
var _hoisted_8 = ["src"];
export function render(_ctx, _cache) {
  var _component_var_icon = _resolveComponent("var-icon");

  var _component_var_form_details = _resolveComponent("var-form-details");

  var _component_var_popup = _resolveComponent("var-popup");

  var _directive_ripple = _resolveDirective("ripple");

  return _openBlock(), _createElementBlock("div", _hoisted_1, [_createElementVNode("div", _hoisted_2, [(_openBlock(true), _createElementBlock(_Fragment, null, _renderList(_ctx.modelValue, f => {
    return _withDirectives((_openBlock(), _createElementBlock("div", {
      class: _normalizeClass(["var-uploader__file var-elevation--2", [f.state === 'loading' ? 'var-uploader--loading' : null]]),
      key: f.id,
      onClick: $event => _ctx.preview(f)
    }, [_createElementVNode("div", _hoisted_4, _toDisplayString(f.name || f.url), 1
    /* TEXT */
    ), _ctx.removable ? (_openBlock(), _createElementBlock("div", {
      key: 0,
      class: "var-uploader__file-close",
      onClick: _withModifiers($event => _ctx.handleRemove(f), ["stop"])
    }, [_createVNode(_component_var_icon, {
      class: "var-uploader__file-close-icon",
      "var-uploader-cover": "",
      name: "delete"
    })], 8
    /* PROPS */
    , _hoisted_5)) : _createCommentVNode("v-if", true), f.cover ? (_openBlock(), _createElementBlock("img", {
      key: 1,
      class: "var-uploader__file-cover",
      style: _normalizeStyle({
        objectFit: f.fit
      }),
      src: f.cover,
      alt: f.name
    }, null, 12
    /* STYLE, PROPS */
    , _hoisted_6)) : _createCommentVNode("v-if", true), _createElementVNode("div", {
      class: _normalizeClass(["var-uploader__file-indicator", [f.state === 'success' ? 'var-uploader--success' : null, f.state === 'error' ? 'var-uploader--error' : null]])
    }, null, 2
    /* CLASS */
    )], 10
    /* CLASS, PROPS */
    , _hoisted_3)), [[_directive_ripple, {
      disabled: _ctx.disabled || _ctx.formDisabled || _ctx.readonly || _ctx.formReadonly || !_ctx.ripple
    }]]);
  }), 128
  /* KEYED_FRAGMENT */
  )), !_ctx.maxlength || _ctx.modelValue.length < _ctx.maxlength ? _withDirectives((_openBlock(), _createElementBlock("div", {
    key: 0,
    class: _normalizeClass(["var--relative", [!_ctx.$slots.default ? 'var-uploader__action var-elevation--2' : null, _ctx.disabled || _ctx.formDisabled ? 'var-uploader--disabled' : null]])
  }, [_createElementVNode("input", {
    class: "var-uploader__action-input",
    type: "file",
    multiple: _ctx.multiple,
    accept: _ctx.accept,
    capture: _ctx.capture,
    disabled: _ctx.disabled || _ctx.formDisabled || _ctx.readonly || _ctx.formReadonly,
    onChange: _cache[0] || (_cache[0] = (...args) => _ctx.handleChange && _ctx.handleChange(...args))
  }, null, 40
  /* PROPS, HYDRATE_EVENTS */
  , _hoisted_7), _renderSlot(_ctx.$slots, "default", {}, () => [_createVNode(_component_var_icon, {
    class: "var-uploader__action-icon",
    "var-uploader-cover": "",
    name: "plus"
  })])], 2
  /* CLASS */
  )), [[_directive_ripple, {
    disabled: _ctx.disabled || _ctx.formDisabled || _ctx.readonly || _ctx.formReadonly || !_ctx.ripple || _ctx.$slots.default
  }]]) : _createCommentVNode("v-if", true)]), _createVNode(_component_var_form_details, {
    "error-message": _ctx.errorMessage,
    "maxlength-text": _ctx.maxlengthText
  }, null, 8
  /* PROPS */
  , ["error-message", "maxlength-text"]), _createVNode(_component_var_popup, {
    class: "var-uploader__preview",
    "var-uploader-cover": "",
    position: "center",
    show: _ctx.showPreview,
    "onUpdate:show": _cache[1] || (_cache[1] = $event => _ctx.showPreview = $event),
    onClosed: _cache[2] || (_cache[2] = $event => _ctx.currentPreview = null)
  }, {
    default: _withCtx(() => {
      var _ctx$currentPreview, _ctx$currentPreview2;

      return [_ctx.currentPreview && _ctx.isHTMLSupportVideo((_ctx$currentPreview = _ctx.currentPreview) == null ? void 0 : _ctx$currentPreview.url) ? (_openBlock(), _createElementBlock("video", {
        key: 0,
        class: "var-uploader__preview-video",
        playsinline: "true",
        "webkit-playsinline": "true",
        "x5-playsinline": "true",
        "x5-video-player-type": "h5",
        "x5-video-player-fullscreen": "false",
        controls: "",
        src: (_ctx$currentPreview2 = _ctx.currentPreview) == null ? void 0 : _ctx$currentPreview2.url
      }, null, 8
      /* PROPS */
      , _hoisted_8)) : _createCommentVNode("v-if", true)];
    }),
    _: 1
    /* STABLE */

  }, 8
  /* PROPS */
  , ["show"])]);
}
export default defineComponent({
  render,
  name: 'VarUploader',
  directives: {
    Ripple
  },
  components: {
    VarIcon,
    VarPopup,
    VarFormDetails
  },
  props,

  setup(props) {
    var showPreview = ref(false);
    var currentPreview = ref(null);
    var maxlengthText = computed(() => {
      var {
        maxlength,
        modelValue: {
          length
        }
      } = props;
      return isNumber(maxlength) ? length + " / " + maxlength : '';
    });
    var {
      form,
      bindForm
    } = useForm();
    var {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      // expose
      resetValidation
    } = useValidation();

    var preview = varFile => {
      var {
        disabled,
        readonly,
        previewed
      } = props;

      if (form != null && form.disabled.value || form != null && form.readonly.value || disabled || readonly || !previewed) {
        return;
      }

      var {
        url
      } = varFile;

      if (isString(url) && isHTMLSupportImage(url)) {
        ImagePreview(url);
        return;
      }

      if (isString(url) && isHTMLSupportVideo(url)) {
        currentPreview.value = varFile;
        showPreview.value = true;
      }
    };

    var createVarFile = file => {
      return {
        id: fid++,
        url: '',
        cover: '',
        name: file.name,
        file
      };
    };

    var getFiles = event => {
      var el = event.target;
      var {
        files: fileList
      } = el;
      return Array.from(fileList);
    };

    var resolver = varFile => {
      return new Promise(resolve => {
        var fileReader = new FileReader();

        fileReader.onload = () => {
          var base64 = fileReader.result;
          varFile.file.type.startsWith('image') && (varFile.cover = base64);
          varFile.url = base64;
          resolve(varFile);
        };

        fileReader.readAsDataURL(varFile.file);
      });
    };

    var getResolvers = varFiles => varFiles.map(resolver);

    var getBeforeReaders = varFiles => {
      var {
        onBeforeRead
      } = props;
      return varFiles.map(varFile => {
        return new Promise(resolve => {
          var valid = onBeforeRead ? onBeforeRead(reactive(varFile)) : true;
          Promise.resolve(valid).then(valid => resolve({
            valid,
            varFile
          }));
        });
      });
    };

    var handleChange = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(function* (event) {
        var _props$onUpdateModel;

        var {
          maxsize,
          maxlength,
          modelValue,
          onOversize,
          onAfterRead,
          readonly,
          disabled
        } = props;

        if (form != null && form.disabled.value || form != null && form.readonly.value || disabled || readonly) {
          return;
        }

        var getValidSizeVarFile = varFiles => {
          return varFiles.filter(varFile => {
            if (varFile.file.size > toNumber(maxsize)) {
              onOversize == null ? void 0 : onOversize(reactive(varFile));
              return false;
            }

            return true;
          });
        };

        var getValidLengthVarFiles = varFiles => {
          var limit = Math.min(varFiles.length, toNumber(maxlength) - modelValue.length);
          return varFiles.slice(0, limit);
        }; // limit


        var files = getFiles(event);
        var varFiles = files.map(createVarFile);
        varFiles = maxsize != null ? getValidSizeVarFile(varFiles) : varFiles;
        varFiles = maxlength != null ? getValidLengthVarFiles(varFiles) : varFiles; // pre resolve

        var resolvedVarFiles = yield Promise.all(getResolvers(varFiles));
        var validationVarFiles = yield Promise.all(getBeforeReaders(resolvedVarFiles));
        var validVarFiles = validationVarFiles.filter(({
          valid
        }) => valid).map(({
          varFile
        }) => varFile);
        (_props$onUpdateModel = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel.call(props, [...modelValue, ...validVarFiles]);
        event.target.value = '';
        validVarFiles.forEach(varFile => onAfterRead == null ? void 0 : onAfterRead(reactive(varFile)));
      });

      return function handleChange(_x) {
        return _ref.apply(this, arguments);
      };
    }();

    var handleRemove = /*#__PURE__*/function () {
      var _ref2 = _asyncToGenerator(function* (removedVarFile) {
        var _props$onUpdateModel2;

        var {
          disabled,
          readonly,
          modelValue,
          onBeforeRemove,
          onRemove
        } = props;

        if (form != null && form.disabled.value || form != null && form.readonly.value || disabled || readonly) {
          return;
        }

        if (onBeforeRemove && !(yield onBeforeRemove(removedVarFile))) {
          return;
        }

        var expectedFiles = modelValue.filter(varFile => varFile !== removedVarFile);
        onRemove == null ? void 0 : onRemove(removedVarFile);
        validateWithTrigger('onRemove');
        (_props$onUpdateModel2 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel2.call(props, expectedFiles);
      });

      return function handleRemove(_x2) {
        return _ref2.apply(this, arguments);
      };
    }(); // expose


    var getSuccess = () => props.modelValue.filter(varFile => varFile.state === 'success'); // expose


    var getError = () => props.modelValue.filter(varFile => varFile.state === 'error'); // expose


    var getLoading = () => props.modelValue.filter(varFile => varFile.state === 'loading');

    var varFileUtils = {
      getSuccess,
      getError,
      getLoading
    };

    var validateWithTrigger = trigger => {
      nextTick(() => {
        var {
          validateTrigger,
          rules,
          modelValue
        } = props;
        vt(validateTrigger, trigger, rules, modelValue, varFileUtils);
      });
    };

    var callReset = false; // expose

    var validate = () => v(props.rules, props.modelValue, varFileUtils); // expose


    var reset = () => {
      var _props$onUpdateModel3;

      callReset = true;
      (_props$onUpdateModel3 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel3.call(props, []);
      resetValidation();
    };

    var uploaderProvider = {
      validate,
      resetValidation,
      reset
    };
    bindForm == null ? void 0 : bindForm(uploaderProvider);
    watch(() => props.modelValue, () => {
      !callReset && validateWithTrigger('onChange');
      callReset = false;
    }, {
      deep: true
    });
    return {
      showPreview,
      currentPreview,
      errorMessage,
      maxlengthText,
      isHTMLSupportVideo,
      isHTMLSupportImage,
      formDisabled: form == null ? void 0 : form.disabled,
      formReadonly: form == null ? void 0 : form.readonly,
      preview,
      handleChange,
      handleRemove,
      getSuccess,
      getError,
      getLoading,
      validate,
      resetValidation,
      reset
    };
  }

});
