// @varlet/ui/lib/divider/Divider.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _elements = require("../utils/elements");

var _shared = require("../utils/shared");

var _props = require("./props");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  key: 0,
  class: "var-divider__text"
};

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-divider var--box", [_ctx.vertical ? 'var-divider--vertical' : null, _ctx.withText ? 'var-divider--with-text' : null, _ctx.isInset ? 'var-divider--inset' : null, _ctx.dashed ? 'var-divider--dashed' : null]]),
    style: (0, _vue.normalizeStyle)(_ctx.style)
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [_ctx.description ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("span", _hoisted_1, (0, _vue.toDisplayString)(_ctx.description), 1
  /* TEXT */
  )) : (0, _vue.createCommentVNode)("v-if", true)])], 6
  /* CLASS, STYLE */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarDivider',
  props: _props.props,

  setup(props, {
    slots
  }) {
    var state = (0, _vue.reactive)({
      withText: false
    });
    var isInset = (0, _vue.computed)(() => (0, _shared.isBool)(props.inset) ? props.inset : true);
    var style = (0, _vue.computed)(() => {
      var {
        inset,
        vertical,
        margin
      } = props;
      var baseStyle = {
        margin
      };
      if ((0, _shared.isBool)(inset) || inset === 0) return _extends({}, baseStyle); // -18px -> -18

      var _inset = (0, _shared.toNumber)(inset); // -18px -> 18px


      var absInsetWithUnit = Math.abs(_inset) + (inset + '').replace(_inset + '', '');
      return vertical ? _extends({}, baseStyle, {
        height: "calc(80% - " + (0, _elements.toSizeUnit)(absInsetWithUnit) + ")"
      }) : _extends({}, baseStyle, {
        width: "calc(100% - " + (0, _elements.toSizeUnit)(absInsetWithUnit) + ")",
        left: _inset > 0 ? (0, _elements.toSizeUnit)(absInsetWithUnit) : (0, _elements.toSizeUnit)(0)
      });
    });

    var checkHasText = () => {
      state.withText = Boolean(slots.default == null ? void 0 : slots.default().length) || Boolean(props.description);
    };

    (0, _vue.onMounted)(() => {
      checkHasText();
    });
    (0, _vue.onUpdated)(() => {
      checkHasText();
    });
    return _extends({}, (0, _vue.toRefs)(state), {
      style,
      isInset
    });
  }

});

exports.default = _default;
