// @varlet/ui/lib/date-picker/src/day-picker-panel.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _dayjs = _interopRequireDefault(require("dayjs"));

var _isSameOrBefore = _interopRequireDefault(require("dayjs/plugin/isSameOrBefore"));

var _isSameOrAfter = _interopRequireDefault(require("dayjs/plugin/isSameOrAfter"));

var _panelHeader = _interopRequireDefault(require("./panel-header.js"));

var _button = _interopRequireDefault(require("../../button"));

var _vue = require("vue");

var _props = require("../props");

var _shared = require("../../utils/shared");

var _locale = require("../../locale");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

_dayjs.default.extend(_isSameOrBefore.default);

_dayjs.default.extend(_isSameOrAfter.default);

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-day-picker__panel"
};
var _hoisted_2 = {
  class: "var-day-picker__content"
};
var _hoisted_3 = {
  class: "var-day-picker__head"
};
var _hoisted_4 = {
  class: "var-day-picker__body"
};

function render(_ctx, _cache) {
  var _component_panel_header = (0, _vue.resolveComponent)("panel-header");

  var _component_var_button = (0, _vue.resolveComponent)("var-button");

  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [(0, _vue.createElementVNode)("div", _hoisted_2, [(0, _vue.createVNode)(_component_panel_header, {
    type: "day",
    date: _ctx.preview,
    disabled: _ctx.panelBtnDisabled,
    onCheckPanel: _ctx.clickMonth,
    onCheckDate: _ctx.checkDate
  }, null, 8
  /* PROPS */
  , ["date", "disabled", "onCheckPanel", "onCheckDate"]), (0, _vue.createVNode)(_vue.Transition, {
    name: _ctx.reverse ? 'var-date-picker-reverse-translatex' : 'var-date-picker-translatex'
  }, {
    default: (0, _vue.withCtx)(() => [((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
      key: _ctx.panelKey
    }, [(0, _vue.createElementVNode)("ul", _hoisted_3, [((0, _vue.openBlock)(true), (0, _vue.createElementBlock)(_vue.Fragment, null, (0, _vue.renderList)(_ctx.sortWeekList, week => {
      return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("li", {
        key: week.index
      }, (0, _vue.toDisplayString)(_ctx.getDayAbbr(week.index)), 1
      /* TEXT */
      );
    }), 128
    /* KEYED_FRAGMENT */
    ))]), (0, _vue.createElementVNode)("ul", _hoisted_4, [((0, _vue.openBlock)(true), (0, _vue.createElementBlock)(_vue.Fragment, null, (0, _vue.renderList)(_ctx.days, (day, index) => {
      return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("li", {
        key: index
      }, [(0, _vue.createVNode)(_component_var_button, (0, _vue.mergeProps)({
        type: "primary",
        class: ["var-day-picker__button", {
          'var-day-picker__button--usable': day > 0
        }],
        "var-day-picker-cover": "",
        round: "",
        ripple: false
      }, _extends({}, _ctx.buttonProps(day)), {
        onClick: $event => _ctx.chooseDay(day)
      }), {
        default: (0, _vue.withCtx)(() => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.filterDay(day)), 1
        /* TEXT */
        )]),
        _: 2
        /* DYNAMIC */

      }, 1040
      /* FULL_PROPS, DYNAMIC_SLOTS */
      , ["class", "onClick"])]);
    }), 128
    /* KEYED_FRAGMENT */
    ))])]))]),
    _: 1
    /* STABLE */

  }, 8
  /* PROPS */
  , ["name"])])]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'DayPickerPanel',
  components: {
    VarButton: _button.default,
    PanelHeader: _panelHeader.default
  },
  props: {
    choose: {
      type: Object,
      required: true
    },
    preview: {
      type: Object,
      required: true
    },
    current: {
      type: String,
      required: true
    },
    clickMonth: {
      type: Function,
      required: true
    },
    componentProps: {
      type: Object,
      required: true
    }
  },
  emits: ['check-preview', 'choose-day'],

  setup(props, {
    emit
  }) {
    var [currentYear, currentMonth, currentDay] = props.current.split('-');
    var days = (0, _vue.ref)([]);
    var reverse = (0, _vue.ref)(false);
    var panelKey = (0, _vue.ref)(0);
    var panelBtnDisabled = (0, _vue.reactive)({
      left: false,
      right: false
    });
    var isCurrent = (0, _vue.computed)(() => props.preview.previewYear === currentYear && props.preview.previewMonth.index === currentMonth);
    var isSame = (0, _vue.computed)(() => props.choose.chooseYear === props.preview.previewYear && props.choose.chooseMonth.index === props.preview.previewMonth.index);
    var sortWeekList = (0, _vue.computed)(() => {
      var index = _props.WEEK_HEADER.findIndex(week => week.index === props.componentProps.firstDayOfWeek);

      if (index === -1 || index === 0) return _props.WEEK_HEADER;
      return _props.WEEK_HEADER.slice(index).concat(_props.WEEK_HEADER.slice(0, index));
    });

    var getDayAbbr = key => {
      var _pack$value$datePicke, _pack$value$datePicke2;

      return (_pack$value$datePicke = (_pack$value$datePicke2 = _locale.pack.value.datePickerWeekDict) == null ? void 0 : _pack$value$datePicke2[key].abbr) != null ? _pack$value$datePicke : '';
    };

    var filterDay = day => day > 0 ? day : '';

    var initDate = () => {
      var {
        preview: {
          previewMonth,
          previewYear
        }
      } = props;
      var monthNum = (0, _dayjs.default)(previewYear + "-" + previewMonth.index).daysInMonth();
      var firstDayToWeek = (0, _dayjs.default)(previewYear + "-" + previewMonth.index + "-01").day();
      var index = sortWeekList.value.findIndex(week => week.index === "" + firstDayToWeek);
      days.value = [...Array(index).fill(-1), ...Array.from(Array(monthNum + 1).keys())].filter(value => value);
    };

    var initHeader = () => {
      var {
        preview: {
          previewYear,
          previewMonth
        },
        componentProps: {
          max,
          min
        }
      } = props;

      if (max) {
        var date = previewYear + "-" + ((0, _shared.toNumber)(previewMonth.index) + 1);
        panelBtnDisabled.right = !(0, _dayjs.default)(date).isSameOrBefore((0, _dayjs.default)(max), 'month');
      }

      if (min) {
        var _date = previewYear + "-" + ((0, _shared.toNumber)(previewMonth.index) - 1);

        panelBtnDisabled.left = !(0, _dayjs.default)(_date).isSameOrAfter((0, _dayjs.default)(min), 'month');
      }
    };

    var inRange = day => {
      var {
        preview: {
          previewYear,
          previewMonth
        },
        componentProps: {
          min,
          max
        }
      } = props;
      var isBeforeMax = true;
      var isAfterMin = true;
      var previewDate = previewYear + "-" + previewMonth.index + "-" + day;
      if (max) isBeforeMax = (0, _dayjs.default)(previewDate).isSameOrBefore((0, _dayjs.default)(max), 'day');
      if (min) isAfterMin = (0, _dayjs.default)(previewDate).isSameOrAfter((0, _dayjs.default)(min), 'day');
      return isBeforeMax && isAfterMin;
    };

    var shouldChoose = val => {
      var {
        choose: {
          chooseDays,
          chooseRangeDay
        },
        componentProps: {
          range
        }
      } = props;
      if (!chooseRangeDay.length) return false;

      if (range) {
        var isBeforeMax = (0, _dayjs.default)(val).isSameOrBefore((0, _dayjs.default)(chooseRangeDay[1]), 'day');
        var isAfterMin = (0, _dayjs.default)(val).isSameOrAfter((0, _dayjs.default)(chooseRangeDay[0]), 'day');
        return isBeforeMax && isAfterMin;
      }

      return chooseDays.includes(val);
    };

    var buttonProps = day => {
      if (day < 0) {
        return {
          text: true,
          outline: false,
          textColor: ''
        };
      }

      var {
        choose: {
          chooseDay
        },
        preview: {
          previewYear,
          previewMonth
        },
        componentProps: {
          allowedDates,
          color,
          multiple,
          range
        }
      } = props;
      var val = previewYear + "-" + previewMonth.index + "-" + day;

      var dayExist = () => {
        if (range || multiple) return shouldChoose(val);
        return (0, _shared.toNumber)(chooseDay) === day && isSame.value;
      };

      var computeDisabled = () => {
        if (!inRange(day)) return true;
        if (!allowedDates) return false;
        return !allowedDates(val);
      };

      var disabled = computeDisabled();

      var computeText = () => {
        if (disabled) return true;
        if (range || multiple) return !shouldChoose(val);
        return !isSame.value || (0, _shared.toNumber)(chooseDay) !== day;
      };

      var computeOutline = () => {
        // 不满足基本条件， 基本条件为当前年、当前月、当前日并且 showCurrent 为true的情况
        if (!(isCurrent.value && (0, _shared.toNumber)(currentDay) === day && props.componentProps.showCurrent)) return false; // 存在着 disabled

        if ((range || multiple || isSame.value) && disabled) return true; // 在选择范围之外

        if (range || multiple) return !shouldChoose(val); // 同一年但是未被选择的情况

        if (isSame.value) return chooseDay !== currentDay;
        return true;
      };

      var textColorOrCover = () => {
        if (disabled) return '';
        if (computeOutline()) return color != null ? color : '';
        if (dayExist()) return '';
        return 'var-date-picker-color-cover';
      };

      var isCover = textColorOrCover().startsWith('var-date-picker');
      return {
        disabled,
        text: computeText(),
        outline: computeOutline(),
        textColor: isCover ? '' : textColorOrCover(),
        'var-date-picker-color-cover': isCover
      };
    };

    var checkDate = checkType => {
      reverse.value = checkType === 'prev';
      panelKey.value += checkType === 'prev' ? -1 : 1;
      emit('check-preview', 'month', checkType);
    };

    var chooseDay = day => {
      emit('choose-day', day);
    };

    (0, _vue.onMounted)(() => {
      initDate();
      initHeader();
    });
    (0, _vue.watch)(() => props.preview, () => {
      initDate();
      initHeader();
    });
    return {
      days,
      reverse,
      panelKey,
      sortWeekList,
      panelBtnDisabled,
      filterDay,
      getDayAbbr,
      checkDate,
      chooseDay,
      buttonProps
    };
  }

});

exports.default = _default;
