// @varlet/ui/lib/image/Image.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _ripple = _interopRequireDefault(require("../ripple"));

var _lazy = _interopRequireDefault(require("../lazy"));

var _vue = require("vue");

var _props = require("./props");

var _elements = require("../utils/elements");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = ["alt", "lazy-error", "lazy-loading"];
var _hoisted_2 = ["alt", "src"];

function render(_ctx, _cache) {
  var _directive_lazy = (0, _vue.resolveDirective)("lazy");

  var _directive_ripple = (0, _vue.resolveDirective)("ripple");

  return (0, _vue.withDirectives)(((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-image var--box", [!_ctx.block ? 'var--inline-block' : null]]),
    style: (0, _vue.normalizeStyle)({
      width: _ctx.toSizeUnit(_ctx.width),
      height: _ctx.toSizeUnit(_ctx.height),
      'border-radius': _ctx.toSizeUnit(_ctx.radius)
    })
  }, [_ctx.lazy ? (0, _vue.withDirectives)(((0, _vue.openBlock)(), (0, _vue.createElementBlock)("img", {
    key: 0,
    class: "var-image__image",
    alt: _ctx.alt,
    "lazy-error": _ctx.error,
    "lazy-loading": _ctx.loading,
    style: (0, _vue.normalizeStyle)({
      objectFit: _ctx.fit
    }),
    onLoad: _cache[0] || (_cache[0] = (...args) => _ctx.handleLoad && _ctx.handleLoad(...args)),
    onError: _cache[1] || (_cache[1] = (...args) => _ctx.handleError && _ctx.handleError(...args)),
    onClick: _cache[2] || (_cache[2] = (...args) => _ctx.onClick && _ctx.onClick(...args))
  }, null, 44
  /* STYLE, PROPS, HYDRATE_EVENTS */
  , _hoisted_1)), [[_directive_lazy, _ctx.src]]) : ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("img", {
    key: 1,
    class: "var-image__image",
    alt: _ctx.alt,
    style: (0, _vue.normalizeStyle)({
      objectFit: _ctx.fit
    }),
    src: _ctx.src,
    onLoad: _cache[3] || (_cache[3] = (...args) => _ctx.handleLoad && _ctx.handleLoad(...args)),
    onError: _cache[4] || (_cache[4] = (...args) => _ctx.handleError && _ctx.handleError(...args)),
    onClick: _cache[5] || (_cache[5] = (...args) => _ctx.onClick && _ctx.onClick(...args))
  }, null, 44
  /* STYLE, PROPS, HYDRATE_EVENTS */
  , _hoisted_2))], 6
  /* CLASS, STYLE */
  )), [[_directive_ripple, {
    disabled: !_ctx.ripple
  }]]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarImage',
  directives: {
    Lazy: _lazy.default,
    Ripple: _ripple.default
  },
  props: _props.props,

  setup(props) {
    var handleLoad = e => {
      var el = e.currentTarget;
      var {
        lazy,
        onLoad,
        onError
      } = props;

      if (lazy) {
        el._lazy.state === 'success' && (onLoad == null ? void 0 : onLoad(e));
        el._lazy.state === 'error' && (onError == null ? void 0 : onError(e));
      } else {
        onLoad == null ? void 0 : onLoad(e);
      }
    };

    var handleError = e => {
      var {
        lazy,
        onError
      } = props;
      !lazy && (onError == null ? void 0 : onError(e));
    };

    return {
      toSizeUnit: _elements.toSizeUnit,
      handleLoad,
      handleError
    };
  }

});

exports.default = _default;
