// @varlet/ui/lib/chip/Chip.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _icon = _interopRequireDefault(require("../icon"));

var _vue = require("vue");

var _props = require("./props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function render(_ctx, _cache) {
  var _component_var_icon = (0, _vue.resolveComponent)("var-icon");

  return (0, _vue.openBlock)(), (0, _vue.createBlock)(_vue.Transition, {
    name: "var-fade"
  }, {
    default: (0, _vue.withCtx)(() => [(0, _vue.createElementVNode)("span", (0, _vue.mergeProps)({
      class: ["var-chip var--box", _ctx.contentClass],
      style: _ctx.chipStyles
    }, _ctx.$attrs), [(0, _vue.renderSlot)(_ctx.$slots, "left"), (0, _vue.createElementVNode)("span", {
      class: (0, _vue.normalizeClass)(["var-chip--text-" + _ctx.size])
    }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 2
    /* CLASS */
    ), (0, _vue.renderSlot)(_ctx.$slots, "right"), _ctx.closable ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("span", {
      key: 0,
      class: "var-chip--close",
      onClick: _cache[0] || (_cache[0] = (...args) => _ctx.onClose && _ctx.onClose(...args))
    }, [(0, _vue.createVNode)(_component_var_icon, {
      name: "" + (_ctx.iconName ? _ctx.iconName : 'close-circle')
    }, null, 8
    /* PROPS */
    , ["name"])])) : (0, _vue.createCommentVNode)("v-if", true)], 16
    /* FULL_PROPS */
    )]),
    _: 3
    /* FORWARDED */

  });
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarChip',
  components: {
    VarIcon: _icon.default
  },
  inheritAttrs: false,
  props: _props.props,

  setup(props) {
    var chipStyles = (0, _vue.computed)(() => {
      var {
        plain,
        textColor,
        color
      } = props;

      if (plain) {
        return {
          color: textColor || color,
          borderColor: color
        };
      }

      return {
        color: textColor,
        background: color
      };
    });
    var contentClass = (0, _vue.computed)(() => {
      var {
        size,
        block,
        type,
        plain,
        round
      } = props;
      var blockClass = block ? 'var--flex' : 'var--inline-flex';
      var plainTypeClass = plain ? "var-chip--plain var-chip--plain-" + type : "var-chip--" + type;
      var roundClass = round && 'var-chip--round';
      return ["var-chip--" + size, blockClass, plainTypeClass, roundClass];
    });
    return {
      chipStyles,
      contentClass
    };
  }

});

exports.default = _default;
