// @varlet/ui/lib/checkbox-group/CheckboxGroup.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _formDetails = _interopRequireDefault(require("../form-details"));

var _vue = require("vue");

var _props = require("./props");

var _components = require("../utils/components");

var _provide = require("./provide");

var _provide2 = require("../form/provide");

var _shared = require("../utils/shared");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-checkbox-group__wrap"
};

function render(_ctx, _cache) {
  var _component_var_form_details = (0, _vue.resolveComponent)("var-form-details");

  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [(0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-checkbox-group", ["var-checkbox-group--" + _ctx.direction]])
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 2
  /* CLASS */
  ), (0, _vue.createVNode)(_component_var_form_details, {
    "error-message": _ctx.errorMessage
  }, null, 8
  /* PROPS */
  , ["error-message"])]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarCheckboxGroup',
  components: {
    VarFormDetails: _formDetails.default
  },
  props: _props.props,

  setup(props) {
    var max = (0, _vue.computed)(() => props.max);
    var checkedCount = (0, _vue.computed)(() => props.modelValue.length);
    var {
      length,
      checkboxes,
      bindCheckboxes
    } = (0, _provide.useCheckboxes)();
    var {
      bindForm
    } = (0, _provide2.useForm)();
    var {
      errorMessage,
      validateWithTrigger: vt,
      validate: v,
      // expose
      resetValidation
    } = (0, _components.useValidation)();
    var checkboxGroupErrorMessage = (0, _vue.computed)(() => errorMessage.value);

    var validateWithTrigger = trigger => {
      (0, _vue.nextTick)(() => {
        var {
          validateTrigger,
          rules,
          modelValue
        } = props;
        vt(validateTrigger, trigger, rules, modelValue);
      });
    };

    var change = changedModelValue => {
      var _props$onUpdateModel;

      (_props$onUpdateModel = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel.call(props, changedModelValue);
      props.onChange == null ? void 0 : props.onChange(changedModelValue);
      validateWithTrigger('onChange');
    };

    var onChecked = changedValue => {
      var {
        modelValue
      } = props;

      if (!modelValue.includes(changedValue)) {
        change([...modelValue, changedValue]);
      }
    };

    var onUnchecked = changedValue => {
      var {
        modelValue
      } = props;

      if (!modelValue.includes(changedValue)) {
        return;
      }

      change(modelValue.filter(value => value !== changedValue));
    };

    var syncCheckboxes = () => checkboxes.forEach(({
      sync
    }) => sync(props.modelValue)); // expose


    var checkAll = () => {
      var _props$onUpdateModel2;

      var checkedValues = checkboxes.map(({
        checkedValue
      }) => checkedValue.value);
      var changedModelValue = (0, _shared.uniq)(checkedValues);
      (_props$onUpdateModel2 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel2.call(props, changedModelValue);
      return changedModelValue;
    }; // expose


    var inverseAll = () => {
      var _props$onUpdateModel3;

      var checkedValues = checkboxes.filter(({
        checked
      }) => !checked.value).map(({
        checkedValue
      }) => checkedValue.value);
      var changedModelValue = (0, _shared.uniq)(checkedValues);
      (_props$onUpdateModel3 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel3.call(props, changedModelValue);
      return changedModelValue;
    }; // expose


    var reset = () => {
      var _props$onUpdateModel4;

      (_props$onUpdateModel4 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel4.call(props, []);
      resetValidation();
    }; // expose


    var validate = () => v(props.rules, props.modelValue);

    (0, _vue.watch)(() => props.modelValue, syncCheckboxes, {
      deep: true
    });
    (0, _vue.watch)(() => length.value, syncCheckboxes);
    var checkboxGroupProvider = {
      max,
      checkedCount,
      onChecked,
      onUnchecked,
      validate,
      resetValidation,
      reset,
      errorMessage: checkboxGroupErrorMessage
    };
    bindCheckboxes(checkboxGroupProvider);
    bindForm == null ? void 0 : bindForm(checkboxGroupProvider);
    return {
      errorMessage,
      checkAll,
      inverseAll,
      reset,
      validate,
      resetValidation
    };
  }

});

exports.default = _default;
