// @varlet/ui/lib/snackbar/Snackbar.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _core = _interopRequireDefault(require("./core"));

var _components = require("../utils/components");

var _props = require("./props");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function render(_ctx, _cache) {
  var _component_var_snackbar_core = (0, _vue.resolveComponent)("var-snackbar-core");

  return (0, _vue.openBlock)(), (0, _vue.createBlock)(_vue.Teleport, {
    to: _ctx.teleport,
    disabled: _ctx.disabled
  }, [(0, _vue.createVNode)(_vue.Transition, {
    name: "var-snackbar-fade",
    onAfterEnter: _ctx.onOpened,
    onAfterLeave: _ctx.onClosed
  }, {
    default: (0, _vue.withCtx)(() => [(0, _vue.createVNode)(_component_var_snackbar_core, (0, _vue.mergeProps)(_ctx.$props, {
      class: "var-snackbar-transition"
    }), {
      action: (0, _vue.withCtx)(() => [(0, _vue.renderSlot)(_ctx.$slots, "action")]),
      default: (0, _vue.withCtx)(() => [(0, _vue.renderSlot)(_ctx.$slots, "default", {}, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.content), 1
      /* TEXT */
      )])]),
      _: 3
      /* FORWARDED */

    }, 16
    /* FULL_PROPS */
    )]),
    _: 3
    /* FORWARDED */

  }, 8
  /* PROPS */
  , ["onAfterEnter", "onAfterLeave"])], 8
  /* PROPS */
  , ["to", "disabled"]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarSnackbar',
  components: {
    VarSnackbarCore: _core.default
  },
  props: _props.props,

  setup() {
    var {
      disabled
    } = (0, _components.useTeleport)();
    return {
      disabled
    };
  }

});

exports.default = _default;
