// @varlet/ui/lib/date-picker/DatePicker.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _dayjs = _interopRequireDefault(require("dayjs"));

var _monthPickerPanel = _interopRequireDefault(require("./src/month-picker-panel.js"));

var _yearPickerPanel = _interopRequireDefault(require("./src/year-picker-panel.js"));

var _dayPickerPanel = _interopRequireDefault(require("./src/day-picker-panel.js"));

var _props = require("./props");

var _shared = require("../utils/shared");

var _locale = require("../locale");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-date-picker-body"
};

function render(_ctx, _cache) {
  var _component_year_picker_panel = (0, _vue.resolveComponent)("year-picker-panel");

  var _component_month_picker_panel = (0, _vue.resolveComponent)("month-picker-panel");

  var _component_day_picker_panel = (0, _vue.resolveComponent)("day-picker-panel");

  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: (0, _vue.normalizeClass)(["var-date-picker", [_ctx.shadow ? 'var-elevation--2' : null]])
  }, [(0, _vue.createElementVNode)("div", {
    class: "var-date-picker-title",
    style: (0, _vue.normalizeStyle)({
      background: _ctx.headerColor || _ctx.color
    })
  }, [(0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-date-picker-title__year", [_ctx.isYearPanel ? 'var-date-picker-title__year--active' : null]]),
    onClick: _cache[0] || (_cache[0] = $event => _ctx.clickEl('year'))
  }, [(0, _vue.renderSlot)(_ctx.$slots, "year", {
    year: _ctx.previewYear
  }, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.previewYear), 1
  /* TEXT */
  )])], 2
  /* CLASS */
  ), (0, _vue.createElementVNode)("div", {
    class: (0, _vue.normalizeClass)(["var-date-picker-title__date", [!_ctx.isYearPanel ? 'var-date-picker-title__date--active' : null, _ctx.range ? 'var-date-picker-title__date--range' : null]]),
    onClick: _cache[1] || (_cache[1] = $event => _ctx.clickEl('date'))
  }, [(0, _vue.createVNode)(_vue.Transition, {
    name: _ctx.multiple ? '' : _ctx.reverse ? 'var-date-picker-reverse-translatey' : 'var-date-picker-translatey'
  }, {
    default: (0, _vue.withCtx)(() => {
      var _ctx$chooseMonth, _ctx$chooseMonth2, _ctx$chooseMonth3;

      return [_ctx.type === 'month' ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
        key: _ctx.chooseYear + ((_ctx$chooseMonth = _ctx.chooseMonth) == null ? void 0 : _ctx$chooseMonth.index)
      }, [_ctx.range ? (0, _vue.renderSlot)(_ctx.$slots, "range", {
        key: 0,
        choose: _ctx.getChoose.chooseRangeMonth
      }, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.getMonthTitle), 1
      /* TEXT */
      )]) : _ctx.multiple ? (0, _vue.renderSlot)(_ctx.$slots, "multiple", {
        key: 1,
        choose: _ctx.getChoose.chooseMonths
      }, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.getMonthTitle), 1
      /* TEXT */
      )]) : (0, _vue.renderSlot)(_ctx.$slots, "month", {
        key: 2,
        month: (_ctx$chooseMonth2 = _ctx.chooseMonth) == null ? void 0 : _ctx$chooseMonth2.index,
        year: _ctx.chooseYear
      }, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.getMonthTitle), 1
      /* TEXT */
      )])])) : ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
        key: _ctx.chooseYear + ((_ctx$chooseMonth3 = _ctx.chooseMonth) == null ? void 0 : _ctx$chooseMonth3.index) + _ctx.chooseDay
      }, [_ctx.range ? (0, _vue.renderSlot)(_ctx.$slots, "range", {
        key: 0,
        choose: _ctx.formatRange
      }, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.getDateTitle), 1
      /* TEXT */
      )]) : _ctx.multiple ? (0, _vue.renderSlot)(_ctx.$slots, "multiple", {
        key: 1,
        choose: _ctx.getChoose.chooseDays
      }, () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.getDateTitle), 1
      /* TEXT */
      )]) : (0, _vue.renderSlot)(_ctx.$slots, "date", (0, _vue.normalizeProps)((0, _vue.mergeProps)({
        key: 2
      }, _ctx.slotProps)), () => [(0, _vue.createTextVNode)((0, _vue.toDisplayString)(_ctx.getDateTitle), 1
      /* TEXT */
      )])]))];
    }),
    _: 3
    /* FORWARDED */

  }, 8
  /* PROPS */
  , ["name"])], 2
  /* CLASS */
  )], 4
  /* STYLE */
  ), (0, _vue.createElementVNode)("div", _hoisted_1, [(0, _vue.createVNode)(_vue.Transition, {
    name: "var-date-picker-panel-fade"
  }, {
    default: (0, _vue.withCtx)(() => [_ctx.isYearPanel ? ((0, _vue.openBlock)(), (0, _vue.createBlock)(_component_year_picker_panel, {
      key: 0,
      "component-props": _ctx.componentProps,
      preview: _ctx.previewYear,
      onChooseYear: _ctx.getChooseYear
    }, null, 8
    /* PROPS */
    , ["component-props", "preview", "onChooseYear"])) : !_ctx.isYearPanel && _ctx.type === 'month' || _ctx.isMonthPanel ? ((0, _vue.openBlock)(), (0, _vue.createBlock)(_component_month_picker_panel, {
      key: 1,
      current: _ctx.currentDate,
      choose: _ctx.getChoose,
      preview: _ctx.getPreview,
      "click-year": () => _ctx.clickEl('year'),
      "component-props": _ctx.componentProps,
      onChooseMonth: _ctx.getChooseMonth,
      onCheckPreview: _ctx.checkPreview
    }, null, 8
    /* PROPS */
    , ["current", "choose", "preview", "click-year", "component-props", "onChooseMonth", "onCheckPreview"])) : !_ctx.isYearPanel && !_ctx.isMonthPanel && _ctx.type === 'date' ? ((0, _vue.openBlock)(), (0, _vue.createBlock)(_component_day_picker_panel, {
      key: 2,
      current: _ctx.currentDate,
      choose: _ctx.getChoose,
      preview: _ctx.getPreview,
      "component-props": _ctx.componentProps,
      "click-month": () => _ctx.clickEl('month'),
      onChooseDay: _ctx.getChooseDay,
      onCheckPreview: _ctx.checkPreview
    }, null, 8
    /* PROPS */
    , ["current", "choose", "preview", "component-props", "click-month", "onChooseDay", "onCheckPreview"])) : (0, _vue.createCommentVNode)("v-if", true)]),
    _: 1
    /* STABLE */

  })])], 2
  /* CLASS */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarDatePicker',
  components: {
    MonthPickerPanel: _monthPickerPanel.default,
    YearPickerPanel: _yearPickerPanel.default,
    DayPickerPanel: _dayPickerPanel.default
  },
  props: _props.props,

  setup(props) {
    var currentDate = (0, _dayjs.default)().format('YYYY-MM-D');
    var [currentYear, currentMonth, currentDay] = currentDate.split('-');

    var monthDes = _props.MONTH_LIST.find(month => month.index === currentMonth);

    var isYearPanel = (0, _vue.ref)(false);
    var isMonthPanel = (0, _vue.ref)(false);
    var rangeDone = (0, _vue.ref)(true);
    var chooseMonth = (0, _vue.ref)(monthDes);
    var chooseYear = (0, _vue.ref)(currentYear);
    var chooseDay = (0, _vue.ref)(currentDay);
    var previewMonth = (0, _vue.ref)(monthDes);
    var previewYear = (0, _vue.ref)(currentYear);
    var reverse = (0, _vue.ref)(false);
    var chooseMonths = (0, _vue.ref)([currentYear + "-" + currentMonth]);
    var chooseDays = (0, _vue.ref)([currentDate]);
    var chooseRangeMonth = (0, _vue.ref)([currentYear + "-" + currentMonth]);
    var chooseRangeDay = (0, _vue.ref)([currentDate]);
    var componentProps = (0, _vue.reactive)({
      allowedDates: props.allowedDates,
      type: props.type,
      color: props.color,
      firstDayOfWeek: props.firstDayOfWeek,
      min: props.min,
      max: props.max,
      showCurrent: props.showCurrent,
      multiple: props.multiple,
      range: props.range
    });
    var getChoose = (0, _vue.computed)(() => ({
      chooseMonth: chooseMonth.value,
      chooseYear: chooseYear.value,
      chooseDay: chooseDay.value,
      chooseMonths: chooseMonths.value,
      chooseDays: chooseDays.value,
      chooseRangeMonth: chooseRangeMonth.value,
      chooseRangeDay: chooseRangeDay.value
    }));
    var getPreview = (0, _vue.computed)(() => ({
      previewMonth: previewMonth.value,
      previewYear: previewYear.value
    }));
    var getMonthTitle = (0, _vue.computed)(() => {
      var _pack$value$datePicke, _pack$value$datePicke2;

      var {
        multiple,
        range
      } = props;
      if (range) return chooseRangeMonth.value[0] + " ~ " + chooseRangeMonth.value[1];
      var monthName = (_pack$value$datePicke = (_pack$value$datePicke2 = _locale.pack.value.datePickerMonthDict) == null ? void 0 : _pack$value$datePicke2[chooseMonth.value.index].name) != null ? _pack$value$datePicke : '';
      return multiple ? "" + chooseMonths.value.length + _locale.pack.value.datePickerSelected : monthName;
    });
    var getDateTitle = (0, _vue.computed)(() => {
      var _pack$value$datePicke3, _pack$value$datePicke4, _pack$value$datePicke5, _pack$value$datePicke6;

      var {
        multiple,
        range
      } = props;

      if (range) {
        chooseRangeDay.value = chooseRangeDay.value.map(date => (0, _dayjs.default)(date).format('YYYY-MM-DD'));
        return chooseRangeDay.value[0] + " ~ " + chooseRangeDay.value[1];
      }

      if (multiple) return "" + chooseDays.value.length + _locale.pack.value.datePickerSelected;
      var weekIndex = (0, _dayjs.default)(chooseYear.value + "-" + chooseMonth.value.index + "-" + chooseDay.value).day();

      var week = _props.WEEK_HEADER.find(value => value.index === "" + weekIndex);

      var weekName = (_pack$value$datePicke3 = (_pack$value$datePicke4 = _locale.pack.value.datePickerWeekDict) == null ? void 0 : _pack$value$datePicke4[week.index].name) != null ? _pack$value$datePicke3 : '';
      var monthName = (_pack$value$datePicke5 = (_pack$value$datePicke6 = _locale.pack.value.datePickerMonthDict) == null ? void 0 : _pack$value$datePicke6[chooseMonth.value.index].name) != null ? _pack$value$datePicke5 : '';
      var showDay = chooseDay.value.padStart(2, '0');
      if (_locale.pack.value.lang === 'zh-CN') return chooseMonth.value.index + "-" + showDay + " " + weekName.slice(0, 3);
      return weekName.slice(0, 3) + ", " + monthName.slice(0, 3) + " " + chooseDay.value;
    });
    var slotProps = (0, _vue.computed)(() => {
      var weekIndex = (0, _dayjs.default)(chooseYear.value + "-" + chooseMonth.value.index + "-" + chooseDay.value).day();
      return {
        week: "" + weekIndex,
        year: chooseYear.value,
        month: chooseMonth.value.index,
        date: chooseDay.value
      };
    });
    var formatRange = (0, _vue.computed)(() => getChoose.value.chooseRangeDay.map(choose => (0, _dayjs.default)(choose).format('YYYY-MM-DD')));
    var isSameYear = (0, _vue.computed)(() => chooseYear.value === previewYear.value);
    var isSameMonth = (0, _vue.computed)(() => chooseMonth.value.index === previewMonth.value.index);

    var clickEl = type => {
      if (type === 'year') isYearPanel.value = true;else if (type === 'month') isMonthPanel.value = true;else {
        isYearPanel.value = false;
        isMonthPanel.value = false;
      }
    };

    var updateRange = (date, type) => {
      var rangeDate = type === 'month' ? chooseRangeMonth : chooseRangeDay;
      rangeDate.value = rangeDone.value ? [date, date] : [rangeDate.value[0], date];
      rangeDone.value = !rangeDone.value;

      if (rangeDone.value) {
        var _props$onUpdateModel;

        var isChangeOrder = (0, _dayjs.default)(rangeDate.value[0]).isAfter(rangeDate.value[1]);

        var _date = isChangeOrder ? [rangeDate.value[1], rangeDate.value[0]] : [...rangeDate.value];

        (_props$onUpdateModel = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel.call(props, _date);
        props.onChange == null ? void 0 : props.onChange(_date);
      }
    };

    var updateMultiple = (date, type) => {
      var _props$onUpdateModel2;

      var multipleDates = type === 'month' ? chooseMonths : chooseDays;
      var formatType = type === 'month' ? 'YYYY-MM' : 'YYYY-MM-DD';
      var formatDates = multipleDates.value.map(date => (0, _dayjs.default)(date).format(formatType));
      var index = formatDates.findIndex(choose => choose === date);
      if (index === -1) formatDates.push(date);else formatDates.splice(index, 1);
      (_props$onUpdateModel2 = props['onUpdate:modelValue']) == null ? void 0 : _props$onUpdateModel2.call(props, formatDates);
      props.onChange == null ? void 0 : props.onChange(formatDates);
    };

    var getReverse = (dateType, date) => {
      if (!isSameYear.value) return chooseYear.value > previewYear.value;
      if (dateType === 'month') return date.index < chooseMonth.value.index;
      return isSameMonth.value ? date < (0, _shared.toNumber)(chooseDay.value) : chooseMonth.value.index > previewMonth.value.index;
    };

    var getChooseDay = day => {
      var {
        readonly,
        range,
        multiple,
        onChange,
        'onUpdate:modelValue': updateModelValue
      } = props;
      if (day < 0 || readonly) return;
      reverse.value = getReverse('day', day);
      var date = previewYear.value + "-" + previewMonth.value.index + "-" + day;
      var formatDate = (0, _dayjs.default)(date).format('YYYY-MM-DD');
      if (range) updateRange(formatDate, 'day');else if (multiple) updateMultiple(formatDate, 'day');else {
        updateModelValue == null ? void 0 : updateModelValue(formatDate);
        onChange == null ? void 0 : onChange(formatDate);
      }
    };

    var getChooseMonth = month => {
      var {
        type,
        readonly,
        range,
        multiple,
        onChange,
        'onUpdate:modelValue': updateModelValue
      } = props;
      reverse.value = getReverse('month', month);

      if (type === 'month' && !readonly) {
        var date = previewYear.value + "-" + month.index;
        if (range) updateRange(date, 'month');else if (multiple) updateMultiple(date, 'month');else {
          updateModelValue == null ? void 0 : updateModelValue(date);
          onChange == null ? void 0 : onChange(date);
        }
      } else {
        previewMonth.value = month;
      }

      isMonthPanel.value = false;
    };

    var getChooseYear = year => {
      previewYear.value = "" + year;
      isYearPanel.value = false;
      isMonthPanel.value = true;
    };

    var checkPreview = (type, checkType) => {
      var changeValue = checkType === 'prev' ? -1 : 1;

      if (type === 'year') {
        previewYear.value = "" + ((0, _shared.toNumber)(previewYear.value) + changeValue);
      } else {
        var checkIndex = (0, _shared.toNumber)(previewMonth.value.index) + changeValue;

        if (checkIndex < 1) {
          previewYear.value = "" + ((0, _shared.toNumber)(previewYear.value) - 1);
          checkIndex = 12;
        }

        if (checkIndex > 12) {
          previewYear.value = "" + ((0, _shared.toNumber)(previewYear.value) + 1);
          checkIndex = 1;
        }

        previewMonth.value = _props.MONTH_LIST.find(month => (0, _shared.toNumber)(month.index) === checkIndex);
      }
    };

    var checkValue = () => {
      if ((props.multiple || props.range) && !(0, _shared.isArray)(props.modelValue)) {
        console.error('[Varlet] DatePicker: type of prop "modelValue" should be an Array');
        return false;
      }

      if (!props.multiple && !props.range && (0, _shared.isArray)(props.modelValue)) {
        console.error('[Varlet] DatePicker: type of prop "modelValue" should be a String');
        return false;
      }

      return true;
    };

    var invalidFormatDate = date => {
      if ((0, _shared.isArray)(date)) return false;

      if (date === undefined || date === 'Invalid Date') {
        console.error('[Varlet] DatePicker: "modelValue" is an Invalid Date');
        return true;
      }

      return false;
    };

    var rangeInit = (value, type) => {
      var rangeDate = type === 'month' ? chooseRangeMonth : chooseRangeDay;
      var formatType = type === 'month' ? 'YYYY-MM' : 'YYYY-MM-D';
      var formatDateList = value.map(choose => (0, _dayjs.default)(choose).format(formatType)).slice(0, 2);
      var isValid = rangeDate.value.some(date => invalidFormatDate(date));
      if (isValid) return;
      rangeDate.value = formatDateList;
      var isChangeOrder = (0, _dayjs.default)(rangeDate.value[0]).isAfter(rangeDate.value[1]);

      if (rangeDate.value.length === 2 && isChangeOrder) {
        rangeDate.value = [rangeDate.value[1], rangeDate.value[0]];
      }
    };

    var multipleInit = (value, type) => {
      var rangeDate = type === 'month' ? chooseMonths : chooseDays;
      var formatType = type === 'month' ? 'YYYY-MM' : 'YYYY-MM-D'; // 需要去重

      var formatDateList = Array.from(new Set(value.map(choose => (0, _dayjs.default)(choose).format(formatType))));
      rangeDate.value = formatDateList.filter(date => date !== 'Invalid Date');
    };

    var dateInit = value => {
      var formatDate = (0, _dayjs.default)(value).format('YYYY-MM-D');
      if (invalidFormatDate(formatDate)) return;
      var [yearValue, monthValue, dayValue] = formatDate.split('-');

      var monthDes = _props.MONTH_LIST.find(month => month.index === monthValue);

      chooseMonth.value = monthDes;
      chooseYear.value = yearValue;
      chooseDay.value = dayValue;
      previewMonth.value = monthDes;
      previewYear.value = yearValue;
    };

    (0, _vue.watch)(() => props.modelValue, value => {
      if (!checkValue() || invalidFormatDate(value)) return;

      if (props.range) {
        if (!(0, _shared.isArray)(value)) return;
        rangeDone.value = value.length !== 1;
        rangeInit(value, props.type);
      } else if (props.multiple) {
        if (!(0, _shared.isArray)(value)) return;
        multipleInit(value, props.type);
      } else {
        dateInit(value);
      }
    }, {
      immediate: true
    });
    return {
      reverse,
      currentDate,
      chooseMonth,
      chooseYear,
      chooseDay,
      previewYear,
      isYearPanel,
      isMonthPanel,
      getMonthTitle,
      getDateTitle,
      getChoose,
      getPreview,
      componentProps,
      slotProps,
      formatRange,
      clickEl,
      getChooseDay,
      getChooseMonth,
      getChooseYear,
      checkPreview
    };
  }

});

exports.default = _default;
