// @varlet/ui/lib/skeleton/Skeleton.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _props = require("./props");

var _elements = require("../utils/elements");

var _shared = require("../utils/shared");

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var--box var-skeleton"
};
var _hoisted_2 = {
  key: 0,
  class: "var-skeleton__data"
};
var _hoisted_3 = {
  key: 1,
  class: "var-skeleton__content"
};

var _hoisted_4 = /*#__PURE__*/_withScopeId(() => /*#__PURE__*/(0, _vue.createElementVNode)("div", {
  class: "var-skeleton--animation"
}, null, -1
/* HOISTED */
));

var _hoisted_5 = [_hoisted_4];
var _hoisted_6 = {
  class: "var-skeleton__article"
};

var _hoisted_7 = /*#__PURE__*/_withScopeId(() => /*#__PURE__*/(0, _vue.createElementVNode)("div", {
  class: "var-skeleton--animation"
}, null, -1
/* HOISTED */
));

var _hoisted_8 = [_hoisted_7];
var _hoisted_9 = {
  class: "var-skeleton__section"
};

var _hoisted_10 = /*#__PURE__*/_withScopeId(() => /*#__PURE__*/(0, _vue.createElementVNode)("div", {
  class: "var-skeleton--animation"
}, null, -1
/* HOISTED */
));

var _hoisted_11 = [_hoisted_10];

var _hoisted_12 = /*#__PURE__*/_withScopeId(() => /*#__PURE__*/(0, _vue.createElementVNode)("div", {
  class: "var-skeleton--animation"
}, null, -1
/* HOISTED */
));

var _hoisted_13 = [_hoisted_12];
var _hoisted_14 = {
  key: 2,
  class: "var-skeleton__fullscreen"
};

var _hoisted_15 = /*#__PURE__*/_withScopeId(() => /*#__PURE__*/(0, _vue.createElementVNode)("div", {
  class: "var-skeleton--animation"
}, null, -1
/* HOISTED */
));

var _hoisted_16 = [_hoisted_15];

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [!_ctx.loading ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_2, [(0, _vue.renderSlot)(_ctx.$slots, "default")])) : (0, _vue.createCommentVNode)("v-if", true), _ctx.loading && !_ctx.fullscreen ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_3, [_ctx.card ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: "var-skeleton__card",
    style: (0, _vue.normalizeStyle)({
      height: _ctx.toSizeUnit(_ctx.cardHeight)
    })
  }, _hoisted_5, 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("div", _hoisted_6, [_ctx.avatar ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: "var-skeleton__avatar",
    style: (0, _vue.normalizeStyle)({
      width: _ctx.toSizeUnit(_ctx.avatarSize),
      height: _ctx.toSizeUnit(_ctx.avatarSize)
    })
  }, _hoisted_8, 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("div", _hoisted_9, [_ctx.title ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    key: 0,
    class: "var-skeleton__title",
    style: (0, _vue.normalizeStyle)({
      width: _ctx.toSizeUnit(_ctx.titleWidth)
    })
  }, _hoisted_11, 4
  /* STYLE */
  )) : (0, _vue.createCommentVNode)("v-if", true), ((0, _vue.openBlock)(true), (0, _vue.createElementBlock)(_vue.Fragment, null, (0, _vue.renderList)(_ctx.toNumber(_ctx.rows), (r, index) => {
    return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
      class: "var-skeleton__row",
      key: r,
      style: (0, _vue.normalizeStyle)({
        width: _ctx.toSizeUnit(_ctx.rowsWidth[index])
      })
    }, _hoisted_13, 4
    /* STYLE */
    );
  }), 128
  /* KEYED_FRAGMENT */
  ))])])])) : (0, _vue.createCommentVNode)("v-if", true), _ctx.loading && _ctx.fullscreen ? ((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_14, _hoisted_16)) : (0, _vue.createCommentVNode)("v-if", true)]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarSkeleton',
  props: _props.props,

  setup() {
    return {
      toSizeUnit: _elements.toSizeUnit,
      toNumber: _shared.toNumber
    };
  }

});

exports.default = _default;
