// @varlet/ui/lib/steps/Steps.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _provide = require("./provide");

var _props = require("./props");

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
    class: "var-steps",
    style: (0, _vue.normalizeStyle)({
      flexDirection: _ctx.direction === 'horizontal' ? 'row' : 'column'
    })
  }, [(0, _vue.renderSlot)(_ctx.$slots, "default")], 4
  /* STYLE */
  );
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarSteps',
  props: _props.props,

  setup(props) {
    var active = (0, _vue.computed)(() => props.active);
    var activeColor = (0, _vue.computed)(() => props.activeColor);
    var inactiveColor = (0, _vue.computed)(() => props.inactiveColor);
    var direction = (0, _vue.computed)(() => props.direction);
    var {
      length,
      bindStep
    } = (0, _provide.useStep)();

    var clickStep = index => {
      props.onClickStep == null ? void 0 : props.onClickStep(index);
    };

    var stepsProvider = {
      active,
      length,
      direction,
      activeColor,
      inactiveColor,
      clickStep
    };
    bindStep(stepsProvider);
  }

});

exports.default = _default;
