// @varlet/ui/lib/form/Form.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _vue = require("vue");

var _props = require("./props");

var _provide = require("./provide");

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-form"
};

function render(_ctx, _cache) {
  return (0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", _hoisted_1, [(0, _vue.renderSlot)(_ctx.$slots, "default")]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarForm',
  props: _props.props,

  setup(props) {
    var disabled = (0, _vue.computed)(() => props.disabled);
    var readonly = (0, _vue.computed)(() => props.readonly);
    var {
      formItems,
      bindFormItems
    } = (0, _provide.useFormItems)(); // expose

    var validate = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator(function* () {
        var res = yield Promise.all(formItems.map(({
          validate
        }) => validate()));
        return res.every(r => r === true);
      });

      return function validate() {
        return _ref.apply(this, arguments);
      };
    }(); // expose


    var reset = () => formItems.forEach(({
      reset
    }) => reset()); // expose


    var resetValidation = () => formItems.forEach(({
      resetValidation
    }) => resetValidation());

    var formProvider = {
      disabled,
      readonly
    };
    bindFormItems(formProvider);
    return {
      validate,
      reset,
      resetValidation
    };
  }

});

exports.default = _default;
