// @varlet/ui/lib/action-sheet/ActionSheet.js
"use strict";

exports.__esModule = true;
exports.render = render;
exports.default = void 0;

var _ripple = _interopRequireDefault(require("../ripple"));

var _popup = _interopRequireDefault(require("../popup"));

var _icon = _interopRequireDefault(require("../icon"));

var _vue = require("vue");

var _props = require("./props");

var _shared = require("../utils/shared");

var _locale = require("../locale");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _withScopeId = n => ((0, _vue.pushScopeId)(""), n = n(), (0, _vue.popScopeId)(), n);

var _hoisted_1 = {
  class: "var-action-sheet__title"
};
var _hoisted_2 = ["onClick"];
var _hoisted_3 = {
  class: "var-action-sheet__action-name"
};

function render(_ctx, _cache) {
  var _component_var_icon = (0, _vue.resolveComponent)("var-icon");

  var _component_var_popup = (0, _vue.resolveComponent)("var-popup");

  var _directive_ripple = (0, _vue.resolveDirective)("ripple");

  return (0, _vue.openBlock)(), (0, _vue.createBlock)(_component_var_popup, (0, _vue.mergeProps)({
    class: "var-action-sheet__popup-radius",
    position: "bottom",
    overlay: _ctx.overlay,
    "overlay-class": _ctx.overlayClass,
    "overlay-style": _ctx.overlayStyle,
    "lock-scroll": _ctx.lockScroll,
    "close-on-click-overlay": _ctx.closeOnClickOverlay,
    teleport: _ctx.teleport,
    show: _ctx.popupShow
  }, {
    'onUpdate:show': value => {
      var _ctx$$props$onUpdate, _ctx$$props;

      return (_ctx$$props$onUpdate = (_ctx$$props = _ctx.$props)['onUpdate:show']) == null ? void 0 : _ctx$$props$onUpdate.call(_ctx$$props, value);
    }
  }, {
    onOpen: _ctx.onOpen,
    onClose: _ctx.onClose,
    onClosed: _ctx.onClosed,
    onOpened: _ctx.onOpened,
    onRouteChange: _ctx.onRouteChange
  }), {
    default: (0, _vue.withCtx)(() => [(0, _vue.createElementVNode)("div", (0, _vue.mergeProps)({
      class: "var-action-sheet var--box"
    }, _ctx.$attrs), [(0, _vue.renderSlot)(_ctx.$slots, "title", {}, () => [(0, _vue.createElementVNode)("div", _hoisted_1, (0, _vue.toDisplayString)(_ctx.dt(_ctx.title, _ctx.pack.actionSheetTitle)), 1
    /* TEXT */
    )]), (0, _vue.renderSlot)(_ctx.$slots, "actions", {}, () => [((0, _vue.openBlock)(true), (0, _vue.createElementBlock)(_vue.Fragment, null, (0, _vue.renderList)(_ctx.actions, action => {
      return (0, _vue.withDirectives)(((0, _vue.openBlock)(), (0, _vue.createElementBlock)("div", {
        class: (0, _vue.normalizeClass)(["var-action-sheet__action-item", [action.className, action.disabled ? 'var-action-sheet--disabled' : null]]),
        key: action.name,
        style: (0, _vue.normalizeStyle)({
          color: action.color
        }),
        onClick: $event => _ctx.handleSelect(action)
      }, [action.icon ? ((0, _vue.openBlock)(), (0, _vue.createBlock)(_component_var_icon, {
        key: 0,
        class: "var-action-sheet__action-icon",
        "var-action-sheet-cover": "",
        name: action.icon,
        size: action.iconSize
      }, null, 8
      /* PROPS */
      , ["name", "size"])) : (0, _vue.createCommentVNode)("v-if", true), (0, _vue.createElementVNode)("div", _hoisted_3, (0, _vue.toDisplayString)(action.name), 1
      /* TEXT */
      )], 14
      /* CLASS, STYLE, PROPS */
      , _hoisted_2)), [[_directive_ripple, {
        disabled: action.disabled
      }]]);
    }), 128
    /* KEYED_FRAGMENT */
    ))])], 16
    /* FULL_PROPS */
    )]),
    _: 3
    /* FORWARDED */

  }, 16
  /* FULL_PROPS */
  , ["overlay", "overlay-class", "overlay-style", "lock-scroll", "close-on-click-overlay", "teleport", "show", "onOpen", "onClose", "onClosed", "onOpened", "onRouteChange"]);
}

var _default = (0, _vue.defineComponent)({
  render,
  name: 'VarActionSheet',
  directives: {
    Ripple: _ripple.default
  },
  components: {
    VarPopup: _popup.default,
    VarIcon: _icon.default
  },
  inheritAttrs: false,
  props: _props.props,

  setup(props) {
    var popupShow = (0, _vue.ref)(false);

    var handleSelect = action => {
      var _props$onUpdateShow;

      if (action.disabled) {
        return;
      }

      var {
        closeOnClickAction,
        onSelect
      } = props;
      onSelect == null ? void 0 : onSelect(action);
      closeOnClickAction && ((_props$onUpdateShow = props['onUpdate:show']) == null ? void 0 : _props$onUpdateShow.call(props, false));
    };

    (0, _vue.watch)(() => props.show, newValue => {
      popupShow.value = newValue;
    }, {
      immediate: true
    });
    return {
      popupShow,
      pack: _locale.pack,
      dt: _shared.dt,
      handleSelect
    };
  }

});

exports.default = _default;
